<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/RDM_NGS_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop a basic intro to RDM, and how to apply good practices to your NGS data",
      "keywords": [
        "RDM, Research Data Management, NGS, Bioinformatics, Data Management Plan"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "# Link to license webpage (https://creativecommons.org/licenses/by-sa/4.0/)",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "# Link to DOI or zenodo materials (10.5281/zenodo.7660225)",
          "@type": "Thing"
        },
        {
          "@id": "#Another link to DOI or zenodo materials (10.5281/zenodo.7565997)",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to NGS Data Management",
      "educationalLevel": "Beginner",
      "audience": "PhD, postdocs, MSc, etc"
    }
    </script><title>File structure and naming conventions - RDM for NGS course</title><link rel=stylesheet href=assets/stylesheets/main.fad675c6.min.css><link rel=stylesheet href=assets/stylesheets/palette.356b1318.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config",""),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#file-structure-and-naming-conventions class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="RDM for NGS course" class="md-header__button md-logo" aria-label="RDM for NGS course" data-md-component=logo> <img src=images/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RDM for NGS course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> File structure and naming conventions </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/RDM_NGS_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> RDM_NGS_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Introduction </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=01_RDM_intro.html class=md-tabs__link> Course contents </a> </li> <li class=md-tabs__item> <a href=practical_workshop.html class=md-tabs__link> Practical workshop </a> </li> <li class=md-tabs__item> <a href=keywords.html class=md-tabs__link> Keyword index </a> </li> <li class=md-tabs__item> <a href=contributors.html class=md-tabs__link> Contributors </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="RDM for NGS course" class="md-nav__button md-logo" aria-label="RDM for NGS course" data-md-component=logo> <img src=images/logo.png alt=logo> </a> RDM for NGS course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/RDM_NGS_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> RDM_NGS_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> <span class=md-ellipsis> Introduction </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex> <span class=md-ellipsis> Course contents </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Course contents </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=01_RDM_intro.html class=md-nav__link> <span class=md-ellipsis> What is RDM </span> </a> </li> <li class=md-nav__item> <a href=02_NGS_data.html class=md-nav__link> <span class=md-ellipsis> NGS data </span> </a> </li> <li class=md-nav__item> <a href=03_data_life_cycle.html class=md-nav__link> <span class=md-ellipsis> Data Life Cycle </span> </a> </li> <li class=md-nav__item> <a href=04_OS_FAIR.html class=md-nav__link> <span class=md-ellipsis> Open Science and FAIR principles </span> </a> </li> <li class=md-nav__item> <a href=05_DMP.html class=md-nav__link> <span class=md-ellipsis> Data Management Plans </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> File structure and naming conventions </span> <span class="md-nav__icon md-icon"></span> </label> <a href=06_file_structure.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> File structure and naming conventions </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#folder-organization class=md-nav__link> <span class=md-ellipsis> Folder organization </span> </a> </li> <li class=md-nav__item> <a href=#assay-folder class=md-nav__link> <span class=md-ellipsis> Assay folder </span> </a> <nav class=md-nav aria-label="Assay folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#assay-id-code-names class=md-nav__link> <span class=md-ellipsis> Assay ID code names </span> </a> </li> <li class=md-nav__item> <a href=#assay-folder-structure class=md-nav__link> <span class=md-ellipsis> Assay folder structure </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#project-folder class=md-nav__link> <span class=md-ellipsis> Project folder </span> </a> <nav class=md-nav aria-label="Project folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#project-folder-structure class=md-nav__link> <span class=md-ellipsis> Project folder structure </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#template-engine class=md-nav__link> <span class=md-ellipsis> Template engine </span> </a> <nav class=md-nav aria-label="Template engine"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#quick-tutorial-on-cookiecutter class=md-nav__link> <span class=md-ellipsis> Quick tutorial on cookiecutter </span> </a> <nav class=md-nav aria-label="Quick tutorial on cookiecutter"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-create-a-folder-template class=md-nav__link> <span class=md-ellipsis> Step 1: Create a Folder Template </span> </a> </li> <li class=md-nav__item> <a href=#step-2-create-cookiecutterjson class=md-nav__link> <span class=md-ellipsis> Step 2: Create cookiecutter.json </span> </a> </li> <li class=md-nav__item> <a href=#step-3-use-cookiecutter class=md-nav__link> <span class=md-ellipsis> Step 3: Use Cookiecutter </span> </a> </li> <li class=md-nav__item> <a href=#step-4-explore-the-generated-project class=md-nav__link> <span class=md-ellipsis> Step 4: Explore the Generated Project </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#genomic-resources-folder class=md-nav__link> <span class=md-ellipsis> Genomic resources folder </span> </a> <nav class=md-nav aria-label="Genomic resources folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#refgenie class=md-nav__link> <span class=md-ellipsis> RefGenie </span> </a> </li> <li class=md-nav__item> <a href=#manual-download class=md-nav__link> <span class=md-ellipsis> Manual download </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#naming-conventions class=md-nav__link> <span class=md-ellipsis> Naming conventions </span> </a> <nav class=md-nav aria-label="Naming conventions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-tips class=md-nav__link> <span class=md-ellipsis> General tips </span> </a> </li> <li class=md-nav__item> <a href=#suggestions-for-ngs-data class=md-nav__link> <span class=md-ellipsis> Suggestions for NGS data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#wrap-up class=md-nav__link> <span class=md-ellipsis> Wrap up </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=07_metadata.html class=md-nav__link> <span class=md-ellipsis> Metadata applied to NGS data </span> </a> </li> <li class=md-nav__item> <a href=08_database.html class=md-nav__link> <span class=md-ellipsis> Database of your data and projects </span> </a> </li> <li class=md-nav__item> <a href=09_version_control.html class=md-nav__link> <span class=md-ellipsis> Version control of your data analysis </span> </a> </li> <li class=md-nav__item> <a href=10_repos.html class=md-nav__link> <span class=md-ellipsis> Archiving your data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=practical_workshop.html class=md-nav__link> <span class=md-ellipsis> Practical workshop </span> </a> </li> <li class=md-nav__item> <a href=keywords.html class=md-nav__link> <span class=md-ellipsis> Keyword index </span> </a> </li> <li class=md-nav__item> <a href=contributors.html class=md-nav__link> <span class=md-ellipsis> Contributors </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#folder-organization class=md-nav__link> <span class=md-ellipsis> Folder organization </span> </a> </li> <li class=md-nav__item> <a href=#assay-folder class=md-nav__link> <span class=md-ellipsis> Assay folder </span> </a> <nav class=md-nav aria-label="Assay folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#assay-id-code-names class=md-nav__link> <span class=md-ellipsis> Assay ID code names </span> </a> </li> <li class=md-nav__item> <a href=#assay-folder-structure class=md-nav__link> <span class=md-ellipsis> Assay folder structure </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#project-folder class=md-nav__link> <span class=md-ellipsis> Project folder </span> </a> <nav class=md-nav aria-label="Project folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#project-folder-structure class=md-nav__link> <span class=md-ellipsis> Project folder structure </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#template-engine class=md-nav__link> <span class=md-ellipsis> Template engine </span> </a> <nav class=md-nav aria-label="Template engine"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#quick-tutorial-on-cookiecutter class=md-nav__link> <span class=md-ellipsis> Quick tutorial on cookiecutter </span> </a> <nav class=md-nav aria-label="Quick tutorial on cookiecutter"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-create-a-folder-template class=md-nav__link> <span class=md-ellipsis> Step 1: Create a Folder Template </span> </a> </li> <li class=md-nav__item> <a href=#step-2-create-cookiecutterjson class=md-nav__link> <span class=md-ellipsis> Step 2: Create cookiecutter.json </span> </a> </li> <li class=md-nav__item> <a href=#step-3-use-cookiecutter class=md-nav__link> <span class=md-ellipsis> Step 3: Use Cookiecutter </span> </a> </li> <li class=md-nav__item> <a href=#step-4-explore-the-generated-project class=md-nav__link> <span class=md-ellipsis> Step 4: Explore the Generated Project </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#genomic-resources-folder class=md-nav__link> <span class=md-ellipsis> Genomic resources folder </span> </a> <nav class=md-nav aria-label="Genomic resources folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#refgenie class=md-nav__link> <span class=md-ellipsis> RefGenie </span> </a> </li> <li class=md-nav__item> <a href=#manual-download class=md-nav__link> <span class=md-ellipsis> Manual download </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#naming-conventions class=md-nav__link> <span class=md-ellipsis> Naming conventions </span> </a> <nav class=md-nav aria-label="Naming conventions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-tips class=md-nav__link> <span class=md-ellipsis> General tips </span> </a> </li> <li class=md-nav__item> <a href=#suggestions-for-ngs-data class=md-nav__link> <span class=md-ellipsis> Suggestions for NGS data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#wrap-up class=md-nav__link> <span class=md-ellipsis> Wrap up </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <nav class=md-tags> <a href=keywords.html#file-structure class=md-tag>File structure</a> <a href=keywords.html#file-organization class=md-tag>File organization</a> <a href=keywords.html#naming-conventions class=md-tag>Naming conventions</a> <a href=keywords.html#folder-templates class=md-tag>Folder templates</a> </nav> <div><h1 id=file-structure-and-naming-conventions>File structure and naming conventions<a class=headerlink href=#file-structure-and-naming-conventions title="Permanent link">¶</a></h1> <p><strong>Last updated:</strong> <em><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 30, 2023</span></em></p> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>🕰 <strong>Time Estimation:</strong> X minutes </p> <p>💬 <strong>Learning Objectives:</strong> </p> <ol> <li>Learn how to organize your NGS data.</li> <li>Separate your data into projects and assays.</li> <li>Manage genomic resources.</li> <li>Learn general rules for naming files and folders.</li> <li>Create your own rules for naming specific results and figures.</li> </ol> </div> <p>We have learn so far how to adhere to FAIR and Open Science standards. You may have noticed that they mostly cover information regarding sharing your data after your project is finished or your data is published. However, how can you <strong>effectively manage your data while you work with it</strong>? How should I organize my data folders, raw data, processed data, data analysis scripts, figures and results? You may not worry too much about it since you will be working on your data all the time, so you will know where everything is. However, projects tend to last long, and the data, its analysis and results, accumulate greatly over time. Not having any sort of structure will incur in many problems in the future, specially once the project is finished or some colleague or collaborator want to access the data. How will they be able to find X or Y table? Will you be able to remember where a figure came from or which script created a csv file? By not applying an efficient and clear system, you are inviting chaos to your doorstep!</p> <p>On the other hand, applying a consistent file structure and naming conventions to your files will help you to efficiently manage your data. Consider the following practices:</p> <ul> <li><strong>Folder structure</strong>: Aim to establish a logical and intuitive folder structure that reflects the organization of research projects and experimental data. Use descriptive folder names to make it easy to locate and access specific data files.<ul> <li><strong>Subfolders</strong>: Use subfolders to further categorize data based on their contents, such as code notebooks, results, reports, etc. This helps to keep data organized and facilitates quick retrieval.</li> </ul> </li> <li><strong>File naming conventions</strong>: implement a standardized file naming convention to ensure consistency and clarity. Use descriptive names that include relevant information, such as type of plots, results tables, etc.</li> </ul> <p>In this lesson we will see a practical example on how you could organize your own files and folders.</p> <h2 id=folder-organization>Folder organization<a class=headerlink href=#folder-organization title="Permanent link">¶</a></h2> <p>You will probably want to divide your NGS data into three different types of folders:</p> <ol> <li><strong>Assay folders</strong>: This folder contains the <strong>raw and processed NGS datasets</strong>, as well as the <strong>pipeline/workflow</strong> used to generate the processed data, provenance of the raw data and quality control reports of the data. This data should be <strong>locked and read-only</strong> to prevent unwanted modifications.</li> <li><strong>Project folders</strong>: This folder contains <strong>all the necessary files for a specific research project</strong>. A project may use several assays or results from other projects. The assay data should not be copied or duplicated, but linked from the original source.</li> <li><strong>Genomic resources folders</strong>: This folder contains <strong>common genomic resources</strong>, such as genome references (fasta files) and annotations (gtf files) for different species, as well as indexes for different alignment algorithms. This data should also be <strong>locked and read-only</strong> to prevent unwanted modifications.</li> </ol> <p>This will help you to keep your data tidied up, specially if you are working on a big lab where assays may be used for different purposes and different people!</p> <h2 id=assay-folder>Assay folder<a class=headerlink href=#assay-folder title="Permanent link">¶</a></h2> <p>For each NGS experiment there should be an <code>Assay</code> folder that will contain all experimental datasets, that is, an <code>Assay</code> (raw files and pipeline processed files). Raw files should not be modified at all, but you should probably lock modifications to the final results once you are done with preprocessing the data. This will help you prevent unwanted modifications to the data. Each <code>Assay</code> subfolder should be named in a way that it is unique, easily readable, distinguishable and understood at a glance. For example, you could name an NGS assay using an acronym for the type of NGS assay (RNAseq, ChIPseq, ATACseq), a keyword that represents a unique descriptive element of that assay, and the date. Like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>&lt;Assay-ID&gt;_&lt;keyword&gt;_YYYYMMDD
</code></pre></div> <p>For example <code>CHIP_Oct4_20230101</code> is a ChIPseq assay made on 1<sup>st</sup> January 2023 with the keyword Oct4, so it is easily identifiable by eye.</p> <h3 id=assay-id-code-names>Assay ID code names<a class=headerlink href=#assay-id-code-names title="Permanent link">¶</a></h3> <p>Below is a example list of different Assay-ID code names. You are welcome to use it and expand it as you wish!</p> <ul> <li><code>CHIP</code>: ChIP-seq</li> <li><code>RNA</code>: RNA-seq</li> <li><code>ATAC</code>: ATAC-seq</li> <li><code>SCR</code>: scRNA-seq</li> <li><code>PROT</code>: Mass Spectrometry Assay</li> <li><code>CAT</code>: Cut&amp;Tag</li> <li><code>CAR</code>: Cut&amp;Run</li> <li><code>RIME</code>: Rapid Immunoprecipitation Mass spectrometry of Endogenous proteins</li> <li>...</li> </ul> <h3 id=assay-folder-structure>Assay folder structure<a class=headerlink href=#assay-folder-structure title="Permanent link">¶</a></h3> <p>Next, let's take a look at a possible folder structure and what kind of files you can find there.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>CHIP_Oct4_20230101/
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>├──<span class=w> </span>README.md
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>├──<span class=w> </span>metadata.yml
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a>├──<span class=w> </span>pipeline.md
<a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>├──<span class=w> </span>processed
<a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a>└──<span class=w> </span>raw
<a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=w>   </span>├──<span class=w> </span>.fastq.gz
<a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a><span class=w>   </span>└──<span class=w> </span>samplesheet.csv
</code></pre></div> <ul> <li><strong>README.md</strong>: Long description of the assay in markdown format. It should contain provenance of the raw NGS data (samples, laboratory protocols used, aim of the assay, etc)</li> <li><strong>metadata.yml</strong>: metadata file for the assay describing different keys and important information regarding that assay (<a href=07_metadata.html>see this lesson</a>).</li> <li><strong>pipeline.md</strong>: description of the pipeline used to process raw data, as well as the commands used to run the pipeline.</li> <li><strong>processed</strong>: folder with results of the preprocessing pipeline. Contents depend on the pipeline used.</li> <li><strong>raw</strong>: folder with the raw data.<ul> <li><em>.fastq.gz</em>:In the case of NGS assays, there should be fastq files.</li> <li><em>samplesheet.csv</em>: file that contains metadata information for the samples. This file is used to run the nf-core pipelines. You can also add extra columns with info regarding the experimental variables and batches so it can be used for downstream analysis as well.</li> </ul> </li> </ul> <p>This folder structure is simple and straight forward, and files will be exactly where you expect them to be! By including a description and metadata file, you will always be able to understand what this experiment was and how it was generated. This information will be <strong>specially useful once you need to submit your data for archiving or write a manuscript</strong>. The <strong>metadata</strong> file will be really handy if you want to <strong>collect information on all your NGS datasets and create a database</strong> (<a href=08_database.html>see this lesson</a>)!</p> <p>Note that the processed folder is not expanded upon. This folder will be very dependant on the workflows/pipelines that you use. We strongly recommend that you use reproducible and community curated pipelines, such as the ones developed by the <a href=https://nf-co.re/ >nf-core community</a>, which have a through documentation on the results they produce. For example, imagine you have run an RNAseq experiment and processed the data using the <a href=https://nf-co.re/rnaseq/3.12.0>nf-core:rnaseq pipeline</a> with the pseudoalignment and quantification option:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>processed/
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>├──<span class=w> </span>fastqc/
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>├──<span class=w> </span>multiqc/
<a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a>├──<span class=w> </span>pipeline_info/
<a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a>├──<span class=w> </span>salmon/
<a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>└──<span class=w> </span>trimgalore/
</code></pre></div> <ul> <li><strong>fastqc</strong>: Quality Control results of the raw fastq files.</li> <li><strong>multiqc</strong>: Full compilation of the Quality Control checks for all your samples for the entire pipeline.</li> <li><strong>pipeline_info</strong>: Information and logs about the pipeline used for each of the steps of the workflow.</li> <li><strong>salmon</strong>: Pseudoalignment and quantification results of the salmon algorithm.</li> <li><strong>trimgalore</strong>: Cleaned fastq files and Quality Control results of the cleaned files.</li> </ul> <p>By using a standardized pipeline, you should be able to always find the preprocessing results where you expect them to be! No need to come up with your own file organization (and document it) or worry that your colleagues (or your future self) won't be able to find the data.</p> <h2 id=project-folder>Project folder<a class=headerlink href=#project-folder title="Permanent link">¶</a></h2> <p>On the other hand, we have the other type of folder called <code>Projects</code>. In this folder you will save a <strong>subfolder for each project</strong> that you (or your lab) works on. Each <code>Project</code> subfolder will contain project information and all the data analysis notebooks and scripts used in that project.</p> <p>Projects and Assays are separated from each other because a project may use one or more assays to answer a scientific question, and assays may be reused several times in different projects. This could be, for example, all the data analysis related to a publication (a RNAseq and a ChIPseq experiment), or a comparison between a previous ATACseq experiment (which was used for a older project) with a new laboratory protocol.</p> <p>As like for an Assay folder, the Project folder should be named in a way that it is unique, easily readable, distinguishable and understood at a glance. For example, you could name it after the main author initials, a keyword that represents a unique descriptive element of that assay, and the date:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>&lt;author_initials&gt;_&lt;keyword&gt;_YYYYMMDD
</code></pre></div> <p>For example, <code>JARH_Oct4_20230101</code>, is a project about the gene Oct4 owned by Jose Alejandro Romero Herrera, created on the 1<sup>st</sup> of January of 2023.</p> <h3 id=project-folder-structure>Project folder structure<a class=headerlink href=#project-folder-structure title="Permanent link">¶</a></h3> <p>Next, let's take a look at a possible folder structure and what kind of files you can find there.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>&lt;author_initials&gt;_&lt;keyword&gt;_YYYYMMDD
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>├──<span class=w> </span>data
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>│<span class=w>  </span>└──<span class=w> </span>&lt;Assay-ID&gt;_&lt;keyword&gt;_YYYYMMDD/
<a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a>├──<span class=w> </span>documents
<a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>│<span class=w>  </span>└──<span class=w> </span>Non-sensitive_NGS_research_project_template.docx
<a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a>├──<span class=w> </span>notebooks
<a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis.rmd
<a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a>├──<span class=w> </span>README.md
<a id=__codelineno-4-9 name=__codelineno-4-9 href=#__codelineno-4-9></a>├──<span class=w> </span>reports
<a id=__codelineno-4-10 name=__codelineno-4-10 href=#__codelineno-4-10></a>│<span class=w>  </span>├──<span class=w> </span>figures
<a id=__codelineno-4-11 name=__codelineno-4-11 href=#__codelineno-4-11></a>│<span class=w>  </span>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis/
<a id=__codelineno-4-12 name=__codelineno-4-12 href=#__codelineno-4-12></a>│<span class=w>  </span>│<span class=w>   </span>└──<span class=w> </span>heatmap_sampleCor_20230102.png
<a id=__codelineno-4-13 name=__codelineno-4-13 href=#__codelineno-4-13></a>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis.html
<a id=__codelineno-4-14 name=__codelineno-4-14 href=#__codelineno-4-14></a>├──<span class=w> </span>requirements.txt
<a id=__codelineno-4-15 name=__codelineno-4-15 href=#__codelineno-4-15></a>├──<span class=w> </span>results
<a id=__codelineno-4-16 name=__codelineno-4-16 href=#__codelineno-4-16></a>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis/
<a id=__codelineno-4-17 name=__codelineno-4-17 href=#__codelineno-4-17></a>│<span class=w>      </span>└──<span class=w> </span>DEA_treat-control_LFC1_p01.tsv
<a id=__codelineno-4-18 name=__codelineno-4-18 href=#__codelineno-4-18></a>├──<span class=w> </span>scripts
<a id=__codelineno-4-19 name=__codelineno-4-19 href=#__codelineno-4-19></a>└──<span class=w> </span>metadata.yml
</code></pre></div> <ul> <li><strong>data</strong>: folder that contains symlinks or shortcuts to where the data is, avoiding copying and modification of original files.</li> <li><strong>documents</strong>: folder containing word documents, slides or pdfs related to the project, such as explanations of the data or project, papers, etc. It also contains your <a href=05_DMP.html>Data Management Plan</a>.<ul> <li><em>Non-sensitive_NGS_research_project_template.docx</em>. This is a pre-filled Data Management Plan based on the Horizon Europe guidelines.</li> </ul> </li> <li><strong>notebooks</strong>: folder containing Jupyter, R markdown or Quarto notebooks with the actual data analysis.</li> <li><strong>README.md</strong>: detailed description of the project in markdown format.</li> <li><strong>reports</strong>: notebooks rendered as html/docx/pdf versions, ideal for sharing with colleagues and also as a formal report of the data analysis procedure.<ul> <li><em>figures</em>: figures produced upon rendering notebooks. The figures will be saved under a subfolder named after the notebook that created them. This is for provenance purposes so we know which notebook created which figures.</li> </ul> </li> <li><strong>requirements.txt</strong>: file explaining what software and libraries/packages and their versions are necessary to reproduce the code.</li> <li><strong>results</strong>: results from the data analysis, such as tables with differentially expressed genes, enrichment results, etc.</li> <li><strong>scripts</strong>: folder containing helper scripts needed to run data analysis or reproduce the work of the folder</li> <li><strong>description.yml</strong>: short description of the project.</li> <li><strong>metadata.yml</strong>: metadata file for the assay describing different keys (<a href=07_metadata.html>see this lesson</a>).</li> </ul> <div class="admonition note"> <p class=admonition-title>Note on the <strong>data</strong> folder</p> <p>What do we mean with shortcuts to the data is? If you have divided your Assays and Projects as suggested, you do not really need to copy or duplicate your data again, simply make a shortcut to the original folder, and you will have access to the data as if it is in the project folder! You can use this command in Linux or MacOS to create a folder softlink:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>ln<span class=w> </span>-s<span class=w> </span>path/to/assays/Assays/&lt;ASSAY_ID&gt;<span class=w> </span>/path/to/projects/Projects/&lt;PROJECT_ID&gt;/data/
</code></pre></div> </div> <div class="admonition note"> <p class=admonition-title>Note on the <strong>notebooks</strong> folder</p> <p>Using annotated notebooks is ideal for reproducibility and readability purposes. Notebooks should be labeled numerically in order they were created, or the order of the data analysis steps, if they are related to each other e.g. <code>00_preprocessing.rmd</code> - <code>01_Differential_Expression_Analysis.rmd</code> - <code>02_Functional_Analysis.rmd</code></p> </div> <div class="admonition note"> <p class=admonition-title>Note on the <strong>results</strong> folder</p> <p>Results from your code notebooks should be saved under this folder. Create a subfolder named after the notebook that created them, so you can always identify which notebook created which results!</p> </div> <!-- **Note**: maybe we can make an environment folder or put inside the scripts folder a Dockerfile that should give you an environment where you can reproduce the results of the folder? --> <h2 id=template-engine>Template engine<a class=headerlink href=#template-engine title="Permanent link">¶</a></h2> <p>You may be thinking right now that doing this kind of setup every time you need to create a new folder would be very tedious, and you would be right! Fortunately, it is very easy to create a folder template using <a href=https://github.com/cookiecutter/cookiecutter>cookiecutter</a>. Cookiecutter is a command-line utility that creates projects from cookiecutters (that is, a template), e.g. creating a Python package project from a Python package project template. Here you can find an example of a cookiecutter folder template directed to <a href=https://github.com/brickmanlab/ngs-template>NGS data</a>, where we have applied the structures explained in the previous sections. You are very welcome to adapt it or modify it to your needs!</p> <p>Now, using only cookiecutter might have some issues. If you update the template, it might be really hard to maintain older folders. By using another tool, <a href=https://github.com/cruft/cruft>cruft</a>, when generating assay and project folders, it allows the user to validate and syncronize old templates with the latest version.</p> <h3 id=quick-tutorial-on-cookiecutter>Quick tutorial on cookiecutter<a class=headerlink href=#quick-tutorial-on-cookiecutter title="Permanent link">¶</a></h3> <p>Creating a Cookiecutter template from scratch involves defining a folder structure, creating a <code>cookiecutter.json</code> file, and specifying the placeholders (keywords) that will be replaced during project generation. Let's walk through the process step by step:</p> <h4 id=step-1-create-a-folder-template>Step 1: Create a Folder Template<a class=headerlink href=#step-1-create-a-folder-template title="Permanent link">¶</a></h4> <p>Start by creating a folder with the structure you want for your template. For example, let's create a simple Python project template:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a>my_template/
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>|-- {{cookiecutter.project_name}}
<a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>|   |-- main.py
<a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a>|-- tests
<a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a>|   |-- test_{{cookiecutter.project_name}}.py
<a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a>|-- README.md
</code></pre></div> <p>In this example, <code>{{cookiecutter.project_name}}</code> is a placeholder that will be replaced with the actual project name when the template is used.</p> <h4 id=step-2-create-cookiecutterjson>Step 2: Create <code>cookiecutter.json</code><a class=headerlink href=#step-2-create-cookiecutterjson title="Permanent link">¶</a></h4> <p>In the root of your template folder, create a file named <code>cookiecutter.json</code>. This file will define the variables (keywords) that users will be prompted to fill in. For our Python project template, it might look like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=p>{</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=w>  </span><span class=nt>"project_name"</span><span class=p>:</span><span class=w> </span><span class=s2>"MyProject"</span><span class=p>,</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=w>  </span><span class=nt>"author_name"</span><span class=p>:</span><span class=w> </span><span class=s2>"Your Name"</span><span class=p>,</span>
<a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=w>  </span><span class=nt>"description"</span><span class=p>:</span><span class=w> </span><span class=s2>"A short description of your project"</span>
<a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=p>}</span>
</code></pre></div> <p>These are the questions users will be asked when generating a project based on your template. The values provided here will be used to replace the corresponding placeholders in the template files.</p> <p>In addition to replacing placeholders in file and directory names, Cookiecutter can also automatically fill in information within the contents of text files. This can be useful for providing default configurations or templates for code files. Let's extend our previous example to include a placeholder inside a text file:</p> <p>First, modify the <code>my_template/main.py</code> file to include a placeholder inside its contents:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># main.py</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=k>def</span> <span class=nf>hello</span><span class=p>():</span>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a>    <span class=nb>print</span><span class=p>(</span><span class=s2>"Hello, {{cookiecutter.project_name}}!"</span><span class=p>)</span>
</code></pre></div> <p>Now, the <code>{{cookiecutter.project_name}}</code> placeholder is inside the <code>main.py</code> file. When you run Cookiecutter, it will automatically replace the placeholders not only in file and directory names but also within the contents of text files. After running Cookiecutter, your generated <code>main.py</code> file might look like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># main.py</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=k>def</span> <span class=nf>hello</span><span class=p>():</span>
<a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a>    <span class=nb>print</span><span class=p>(</span><span class=s2>"Hello, MyProject!"</span><span class=p>)</span>  <span class=c1># Assuming "MyProject" was entered as the project_name</span>
</code></pre></div> <h4 id=step-3-use-cookiecutter>Step 3: Use Cookiecutter<a class=headerlink href=#step-3-use-cookiecutter title="Permanent link">¶</a></h4> <p>Now that your template is set up, you can use Cookiecutter to generate a project based on it. Open a terminal and run:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a>cookiecutter<span class=w> </span>path/to/your/template
</code></pre></div> <p>Cookiecutter will prompt you to fill in the values for <code>project_name</code>, <code>author_name</code>, and <code>description</code>. After you provide these values, Cookiecutter will replace the placeholders in your template files with the entered values.</p> <h4 id=step-4-explore-the-generated-project>Step 4: Explore the Generated Project<a class=headerlink href=#step-4-explore-the-generated-project title="Permanent link">¶</a></h4> <p>Once the generation process is complete, navigate to the directory where Cookiecutter created the new project. You will see a project structure with the placeholders replaced by the values you provided.</p> <div class="admonition question"> <p class=admonition-title>Exercise: Create your own template</p> <p>Using cookiecutter, create your own templates for your folders. You do not need to copy exactly our suggestions, adjust your template to your own needs! In order to create your cookiecutter template, your will need to install python, <a href=https://cookiecutter.readthedocs.io/en/stable/installation.html>cookiecutter</a>, Git and a GitHub account. If you do not have Git and a GitHub account, we suggest you do one as soon as possible. We will take a deeper look at Git and GitHub in the <a href=09_version_control.html>version control lesson</a>.</p> <p>We have prepared already two simple cookiecutter templates in GitHub repositories.</p> <p><strong>Assay</strong></p> <ol> <li>First, fork our <code>https://github.com/hds-sandbox/assay-template</code> from the GitHub page into your own account/organization. <img alt=fork_repo_example src=images/fork_repo.png></li> <li>Then, use <code>git clone &lt;your URL to the template&gt;</code> to put it in your computer.</li> <li>Modify the contents of the repository so that it matches the <strong>Assay</strong> example above. You are welcome to do changes as you please!</li> <li>Modify the <code>cookiecutter.json</code> file so that it will include the <strong>Assay</strong> name template</li> <li>Git add, commit and push your changes</li> <li>Test your folder by using <code>cookiecutter &lt;URL to your GitHub repository for "assay-template&gt;</code></li> </ol> <p><strong>Project</strong></p> <ol> <li>First, fork our <code>https://github.com/hds-sandbox/project-template</code> from the GitHub page into your own account/organization. <img alt=fork_repo_example src=images/fork_repo_project.png></li> <li>Then, use <code>git clone &lt;your URL to the template&gt;</code> to put it in your computer.</li> <li>Modify the contents of the repository so that it matches the <strong>Project</strong> example above. You are welcome to do changes as you please!</li> <li>Modify the <code>cookiecutter.json</code> file so that it will include the <strong>Project</strong> name template</li> <li>Git add, commit and push your changes</li> <li>Test your folder by using <code>cookiecutter &lt;URL to your GitHub repository for "project-template&gt;</code></li> </ol> </div> <h2 id=genomic-resources-folder>Genomic resources folder<a class=headerlink href=#genomic-resources-folder title="Permanent link">¶</a></h2> <p>Preprocessing NGS data usually requires different genomic resources in order to align and and annotate fastq files. First and most importantly, you will need reference genomes (human and mouse are the most common ones) in FASTA format. In addition, aligner tools such as STAR, Bowtie, etc., require indexed fasta files to perform alignment of reads. Moreover, GTF or GFF files are necessary in order to quantify reads into genomic regions (such as genes or promoters). Nonetheless, these genomic resources are often updated and are released periodically under different versions from different sources. In order to make your data reproducible, you will need to control and manage these genomic resources and specify their versions and sources. For example, the latest mouse reference genome is GRCm39, but may studies still align their reads to the version GRCm38.</p> <p>How do you keep track of your resources? You could, again, set up a folder structure that hosts reference genomes, annotations and indexes per species and per version, or use a reference genome manager such as <a href=http://refgenie.databio.org/en/latest/ >refgenie</a>. We will take a look at both options.</p> <h3 id=refgenie>RefGenie<a class=headerlink href=#refgenie title="Permanent link">¶</a></h3> <p><a href=http://refgenie.databio.org/en/latest/ >Refgenie</a> manages storage, access, and transfer of reference genome resources. It provides command-line and Python interfaces to download pre-built reference genome "assets", like indexes used by bioinformatics tools. It can also build assets for custom genome assemblies. Refgenie provides programmatic access to a standard genome folder structure, so software can swap from one genome to another.</p> <p>Check <a href=http://refgenie.databio.org/en/latest/tutorial/ >their tutorial</a> if you want to start managing your genomic resources with Refgenie!</p> <h3 id=manual-download>Manual download<a class=headerlink href=#manual-download title="Permanent link">¶</a></h3> <p>If you want to keep manual track and manage yourself and not depend on a third party tool, you can always download the genomic resources directly from the source. We will show you how in this section:</p> <ol> <li>Create a new folder, called <code>genomic_resources</code>.</li> <li>Inside this folder, separate each species with a subfolder.</li> <li>Inside the species subfolder, separate it by the resource version</li> <li>Inside the version subfolder, separate genomic sequences (FASTA) and annotations (GTF/GFF) from different alignment indexes</li> </ol> <p>In the example below, we are downloading human (version GRCh38) and mouse (version GRCm39) genomic resources from the <a href=ftp://ftp.ensembl.org>ensembl FTP server</a></p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a>genomic_resources/
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a>├──<span class=w> </span>homo_sapiens/
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>│<span class=w>  </span>└──<span class=w> </span>GRCh38/
<a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a>│<span class=w>     </span>├──<span class=w> </span>Homo_sapiens.GRCh38.109.gtf.gz
<a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a>│<span class=w>     </span>├──<span class=w> </span>Homo_sapiens.GRCh38.109.dna_sm.primary_assembly.fa.gz
<a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a>│<span class=w>     </span>└──<span class=w> </span>indexes/
<a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a>│<span class=w>        </span>├──<span class=w> </span>salmon/
<a id=__codelineno-11-8 name=__codelineno-11-8 href=#__codelineno-11-8></a>│<span class=w>        </span>└──<span class=w> </span>STAR/
<a id=__codelineno-11-9 name=__codelineno-11-9 href=#__codelineno-11-9></a>├──<span class=w> </span>mus_musculus/
<a id=__codelineno-11-10 name=__codelineno-11-10 href=#__codelineno-11-10></a>│<span class=w>  </span>└──<span class=w> </span>GRCm39/
<a id=__codelineno-11-11 name=__codelineno-11-11 href=#__codelineno-11-11></a>│<span class=w>     </span>├──<span class=w> </span>Mus_musculus.GRCm39.109.gtf.gz
<a id=__codelineno-11-12 name=__codelineno-11-12 href=#__codelineno-11-12></a>│<span class=w>     </span>├──<span class=w> </span>Mus_musculus.GRCm39.109.dna_sm.primary_assembly.fa.gz
<a id=__codelineno-11-13 name=__codelineno-11-13 href=#__codelineno-11-13></a>│<span class=w>     </span>└──<span class=w> </span>indexes/
<a id=__codelineno-11-14 name=__codelineno-11-14 href=#__codelineno-11-14></a>│<span class=w>        </span>├──<span class=w> </span>salmon/
<a id=__codelineno-11-15 name=__codelineno-11-15 href=#__codelineno-11-15></a>│<span class=w>        </span>└──<span class=w> </span>STAR/
<a id=__codelineno-11-16 name=__codelineno-11-16 href=#__codelineno-11-16></a>├──<span class=w> </span>ref_genomes.sh
<a id=__codelineno-11-17 name=__codelineno-11-17 href=#__codelineno-11-17></a>└──<span class=w> </span>create_indexes.sh
</code></pre></div> <p>As you can see, we have also keep here a bash script that downloads and sets up the folders. The contents of the bash script should also indicate a date of when was the data downloaded. For example:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=ch>#!/bin/bash</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a>
<a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=c1># Download human (GRCh38) and mouse (GRCm39) genome and annotations</span>
<a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=c1># Release 109</span>
<a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=c1># 2023-07-10</span>
<a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a>
<a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a><span class=c1># Go to home</span>
<a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a><span class=nb>cd</span>
<a id=__codelineno-12-9 name=__codelineno-12-9 href=#__codelineno-12-9></a><span class=c1># Create homo_sapiens version folder and go to that folder</span>
<a id=__codelineno-12-10 name=__codelineno-12-10 href=#__codelineno-12-10></a>mkdir<span class=w> </span>-p<span class=w> </span>genomic_resources/homo_sapiens/GRCh38
<a id=__codelineno-12-11 name=__codelineno-12-11 href=#__codelineno-12-11></a><span class=nb>cd</span><span class=w>  </span>genomic_resources/homo_sapiens/GRCh38
<a id=__codelineno-12-12 name=__codelineno-12-12 href=#__codelineno-12-12></a>
<a id=__codelineno-12-13 name=__codelineno-12-13 href=#__codelineno-12-13></a><span class=c1># Download FASTA file and GTF file from the ensembl release 109</span>
<a id=__codelineno-12-14 name=__codelineno-12-14 href=#__codelineno-12-14></a>wget<span class=w> </span>-L<span class=w> </span>ftp://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz
<a id=__codelineno-12-15 name=__codelineno-12-15 href=#__codelineno-12-15></a>wget<span class=w> </span>-L<span class=w> </span>ftp://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz
<a id=__codelineno-12-16 name=__codelineno-12-16 href=#__codelineno-12-16></a>
<a id=__codelineno-12-17 name=__codelineno-12-17 href=#__codelineno-12-17></a><span class=c1># Go to home</span>
<a id=__codelineno-12-18 name=__codelineno-12-18 href=#__codelineno-12-18></a><span class=nb>cd</span>
<a id=__codelineno-12-19 name=__codelineno-12-19 href=#__codelineno-12-19></a><span class=c1># Create mus_musculus version folder and go to that folder</span>
<a id=__codelineno-12-20 name=__codelineno-12-20 href=#__codelineno-12-20></a>mkdir<span class=w> </span>-p<span class=w> </span>genomic_resources/mouse/GRCm39
<a id=__codelineno-12-21 name=__codelineno-12-21 href=#__codelineno-12-21></a><span class=nb>cd</span><span class=w>  </span>genomic_resources/mouse/GRCm39
<a id=__codelineno-12-22 name=__codelineno-12-22 href=#__codelineno-12-22></a>
<a id=__codelineno-12-23 name=__codelineno-12-23 href=#__codelineno-12-23></a><span class=c1># Download FASTA file and GTF file from the ensembl release 109</span>
<a id=__codelineno-12-24 name=__codelineno-12-24 href=#__codelineno-12-24></a>wget<span class=w> </span>-L<span class=w> </span>ftp://ftp.ensembl.org/pub/release-109/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa.gz
<a id=__codelineno-12-25 name=__codelineno-12-25 href=#__codelineno-12-25></a>wget<span class=w> </span>-L<span class=w> </span>https://ftp.ensembl.org/pub/release-109/gtf/mus_musculus/Mus_musculus.GRCm39.109.gtf.gz
</code></pre></div> <p>After that is all set up, you can run another bash script that will create a index folder inside each version subfolder and generate indexes for different alignment tools!</p> <h2 id=naming-conventions>Naming conventions<a class=headerlink href=#naming-conventions title="Permanent link">¶</a></h2> <p>Using consistent naming conventions is important in scientific research as it helps with the organization and retrieval of data or results. By adopting standardized naming conventions, researchers ensure that files, experiments, or data sets are labeled in a clear, logical manner. This makes it easier to locate and compare similar types of data or results, even when dealing with large datasets or multiple experiments. For instance, in genomics, employing uniform naming conventions for files related to specific experiments or samples allows for swift identification and comparison of relevant data, streamlining the research process and contributing to the reproducibility of findings. This practice promotes efficiency, collaboration, and the integrity of scientific work.</p> <h3 id=general-tips>General tips<a class=headerlink href=#general-tips title="Permanent link">¶</a></h3> <p>Below you will find a small list of general tips to follow when you name a folder or a file:</p> <ul> <li>Use only alphanumeric characters to write a word: a to z and 0 to 9</li> <li>Avoid special characters: ~!@#$%^&amp;*()`[]{}"|</li> <li>Date format: use <code>YYYYMMDD</code> format. For example: 20230101.</li> <li>Authors: use initials. For example: JARH</li> <li><strong>Don't use of spaces</strong>! Computers get very confused when you need to point a path to a file and it contains spaces! Instead:<ul> <li>Separate field sections are separated by underscores <code>_</code>.</li> <li>Words in each section are written in <a href=https://en.wikipedia.org/wiki/Camel_case>camelCase</a>. It would look then like this: <code>field1_word1Word2.txt</code>. For example: <code>heatmap_sampleCor_20230101.png</code>. The first field indicates what this file is, i.e., a heatmap. Second field is what is being plotted, i.e, sample correlations; since the field contains two words, they are written in camelCase. The third field is the date of when the image was created.</li> </ul> </li> <li>Use as short fields as possible. You can try to use understandable abbreviations, like LFC for LogFoldChange, Cor for correlations, Dist for distances, etc.</li> <li>Avoid long names as much as you can, be concise!</li> <li>Avoid creating many sublevels of folders.</li> <li>Write down your naming convention pattern and document it in the README file</li> <li>When using a sequential numbering system, use leading zeros to make sure files sort in sequential order. Using <code>01</code> instead of just <code>1</code> if your sequence only goes up to <code>99</code>.</li> <li>Versions should be used as the last element, and use at least two digits with a leading 0 (e.g. v01, v02)</li> </ul> <h3 id=suggestions-for-ngs-data>Suggestions for NGS data<a class=headerlink href=#suggestions-for-ngs-data title="Permanent link">¶</a></h3> <p>More info on naming conventions for different types of files and analysis is in development.</p> <table> <thead> <tr> <th style="text-align: left;">name</th> <th style="text-align: left;">description</th> <th style="text-align: left;">naming_convention</th> <th style="text-align: left;">file format</th> <th style="text-align: left;">example</th> </tr> </thead> <tbody> <tr> <td style="text-align: left;">.fastq</td> <td style="text-align: left;">raw sequencing reads</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">sampleID_run_read1.fastq</td> </tr> <tr> <td style="text-align: left;">.fastqc</td> <td style="text-align: left;">quality control from fastqc</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">sampleID_run_read1.fastqc</td> </tr> <tr> <td style="text-align: left;">.bam</td> <td style="text-align: left;">aligned reads</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">sampleID_run_read1.bam</td> </tr> <tr> <td style="text-align: left;">GTF</td> <td style="text-align: left;">sequence annotation</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">one of https://www.gencodegenes.org/</td> </tr> <tr> <td style="text-align: left;">GFF</td> <td style="text-align: left;">sequence annotation</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">one of https://www.gencodegenes.org/</td> </tr> <tr> <td style="text-align: left;">.bed</td> <td style="text-align: left;">genome locations</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">.bigwig</td> <td style="text-align: left;">genome coverage</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">.fasta</td> <td style="text-align: left;">sequence data (nucleotide/aminoacid)</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">one of https://www.gencodegenes.org/</td> </tr> <tr> <td style="text-align: left;">Multiqc report</td> <td style="text-align: left;">QC aggregated report</td> <td style="text-align: left;">&lt;assayID&gt;_YYYYMMDD.multiqc</td> <td style="text-align: left;">multiqc</td> <td style="text-align: left;">RNA_20200101.multiqc</td> </tr> <tr> <td style="text-align: left;">Count matrix</td> <td style="text-align: left;">final count matrix</td> <td style="text-align: left;">&lt;assayID&gt;_cm_aligner_YYYYMMDD.tsv</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">RNA_cm_salmon_20200101.tsv</td> </tr> <tr> <td style="text-align: left;">DEA</td> <td style="text-align: left;">differential expression analysis results</td> <td style="text-align: left;">DEA_&lt;condition1-condition2&gt;_LFC&lt;absolute_threshold&gt;_p&lt;pvalue decimals&gt;_YYYYMMDD.tsv</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">DEA_treat-untreat_LFC1_p01_20200101.tsv</td> </tr> <tr> <td style="text-align: left;">DBA</td> <td style="text-align: left;">differential binding analysis results</td> <td style="text-align: left;">DBA_&lt;condition1-condition2&gt;_LFC&lt;absolute_threshold&gt;_p&lt;pvalue decimals&gt;_YYYYMMDD.tsv</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">DBA_treat-untreat_LFC1_p01_20200101.tsv</td> </tr> <tr> <td style="text-align: left;">MAplot</td> <td style="text-align: left;">MA plot</td> <td style="text-align: left;">MAplot_&lt;condition1-condition2&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">MAplot_treat-untreat_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Heatmap plot</td> <td style="text-align: left;">Heatmap plot of anything</td> <td style="text-align: left;">heatmap_&lt;type&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">heatmap_sampleCor_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Volcano plot</td> <td style="text-align: left;">Volcano plot</td> <td style="text-align: left;">volcano_&lt;condition1-condition2&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">volcano_treat-untreat_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Venn diagram</td> <td style="text-align: left;">Venn diagram</td> <td style="text-align: left;">venn_&lt;type&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">venn_consensus_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Enrichment table</td> <td style="text-align: left;">Enrichment results</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">nan</td> </tr> </tbody> </table> <div class="admonition question"> <p class=admonition-title>Exercise: Create your own naming conventions</p> <p>Think about the most common types of files and folders you will be working on, such as visualizations, results tables, processed files, etc. Then come up with a logical and clear way of naming those files using the tips suggested above. Remember to avoid making long and complicated names!</p> </div> <h2 id=wrap-up>Wrap up<a class=headerlink href=#wrap-up title="Permanent link">¶</a></h2> <p>In this lesson, we have learned some practical tips and examples about how to organize your data and bring some order to chaos! We have shown how to use <code>cookiecutter</code> as a template engine that can be used to create new <code>Assay</code> and <code>Project</code> folders without hassle. Hopefully you are now able to create your own templates and reuse them as much as you need. In the next lesson, we will look at what kind of metadata you should record for your <code>Assay</code> and <code>Project</code> folders, from general fields such as author or date, to more NGS specific like organism, cell type or tissue.</p></div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 30, 2023</span> <br> Created: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 30, 2023</span> </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=05_DMP.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Data Management Plans"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Data Management Plans </div> </div> </a> <a href=07_metadata.html class="md-footer__link md-footer__link--next" aria-label="Next: Metadata applied to NGS data"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Metadata applied to NGS data </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.cd18aaf1.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>