<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/RDM_NGS_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop a basic intro to RDM, and how to apply good practices to your NGS data",
      "keywords": [
        "RDM, Research Data Management, NGS, Bioinformatics, Data Management Plan"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "# Link to license webpage (https://creativecommons.org/licenses/by-sa/4.0/)",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "# Link to DOI or zenodo materials (10.5281/zenodo.7660225)",
          "@type": "Thing"
        },
        {
          "@id": "#Another link to DOI or zenodo materials (10.5281/zenodo.7565997)",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to NGS Data Management",
      "educationalLevel": "Beginner",
      "audience": "PhD, postdocs, MSc, etc"
    }
    </script><title>DTU workshop 2023 - RDM for NGS course</title><link rel=stylesheet href=assets/stylesheets/main.fad675c6.min.css><link rel=stylesheet href=assets/stylesheets/palette.356b1318.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config",""),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#dtu-workshop-october-2023 class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="RDM for NGS course" class="md-header__button md-logo" aria-label="RDM for NGS course" data-md-component=logo> <img src=images/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RDM for NGS course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> DTU workshop 2023 </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/RDM_NGS_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> RDM_NGS_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Introduction </a> </li> <li class=md-tabs__item> <a href=01_RDM_intro.html class=md-tabs__link> Course contents </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=practical_workshop.html class=md-tabs__link> Practical workshop </a> </li> <li class=md-tabs__item> <a href=keywords.html class=md-tabs__link> Keyword index </a> </li> <li class=md-tabs__item> <a href=contributors.html class=md-tabs__link> Contributors </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation hidden> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="RDM for NGS course" class="md-nav__button md-logo" aria-label="RDM for NGS course" data-md-component=logo> <img src=images/logo.png alt=logo> </a> RDM for NGS course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/RDM_NGS_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> RDM_NGS_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> <span class=md-ellipsis> Introduction </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex> <span class=md-ellipsis> Course contents </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Course contents </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=01_RDM_intro.html class=md-nav__link> <span class=md-ellipsis> What is RDM </span> </a> </li> <li class=md-nav__item> <a href=02_NGS_data.html class=md-nav__link> <span class=md-ellipsis> NGS data </span> </a> </li> <li class=md-nav__item> <a href=03_data_life_cycle.html class=md-nav__link> <span class=md-ellipsis> Data Life Cycle </span> </a> </li> <li class=md-nav__item> <a href=04_OS_FAIR.html class=md-nav__link> <span class=md-ellipsis> Open Science and FAIR principles </span> </a> </li> <li class=md-nav__item> <a href=05_DMP.html class=md-nav__link> <span class=md-ellipsis> Data Management Plans </span> </a> </li> <li class=md-nav__item> <a href=06_file_structure.html class=md-nav__link> <span class=md-ellipsis> File structure and naming conventions </span> </a> </li> <li class=md-nav__item> <a href=07_metadata.html class=md-nav__link> <span class=md-ellipsis> Metadata applied to NGS data </span> </a> </li> <li class=md-nav__item> <a href=08_database.html class=md-nav__link> <span class=md-ellipsis> Database of your data and projects </span> </a> </li> <li class=md-nav__item> <a href=09_version_control.html class=md-nav__link> <span class=md-ellipsis> Version control of your data analysis </span> </a> </li> <li class=md-nav__item> <a href=10_repos.html class=md-nav__link> <span class=md-ellipsis> Archiving your data </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Practical workshop </span> <span class="md-nav__icon md-icon"></span> </label> <a href=practical_workshop.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Practical workshop </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-organize-and-structure-your-ngs-data-and-data-analysis class=md-nav__link> <span class=md-ellipsis> 1. Organize and structure your NGS data and data analysis </span> </a> <nav class=md-nav aria-label="1. Organize and structure your NGS data and data analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#assay-folder class=md-nav__link> <span class=md-ellipsis> Assay folder </span> </a> </li> <li class=md-nav__item> <a href=#project-folder class=md-nav__link> <span class=md-ellipsis> Project folder </span> </a> </li> <li class=md-nav__item> <a href=#template-engine class=md-nav__link> <span class=md-ellipsis> Template engine </span> </a> <nav class=md-nav aria-label="Template engine"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#quick-tutorial-on-cookiecutter class=md-nav__link> <span class=md-ellipsis> Quick tutorial on cookiecutter </span> </a> <nav class=md-nav aria-label="Quick tutorial on cookiecutter"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-create-a-folder-template class=md-nav__link> <span class=md-ellipsis> Step 1: Create a Folder Template </span> </a> </li> <li class=md-nav__item> <a href=#step-2-create-cookiecutterjson class=md-nav__link> <span class=md-ellipsis> Step 2: Create cookiecutter.json </span> </a> </li> <li class=md-nav__item> <a href=#step-3-use-cookiecutter class=md-nav__link> <span class=md-ellipsis> Step 3: Use Cookiecutter </span> </a> </li> <li class=md-nav__item> <a href=#step-4-explore-the-generated-project class=md-nav__link> <span class=md-ellipsis> Step 4: Explore the Generated Project </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#2-metadata-and-naming-conventions class=md-nav__link> <span class=md-ellipsis> 2. Metadata and naming conventions </span> </a> <nav class=md-nav aria-label="2. Metadata and naming conventions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#readmemd-file class=md-nav__link> <span class=md-ellipsis> README.md file </span> </a> </li> <li class=md-nav__item> <a href=#metadatayml class=md-nav__link> <span class=md-ellipsis> metadata.yml </span> </a> </li> <li class=md-nav__item> <a href=#metadata-fields class=md-nav__link> <span class=md-ellipsis> Metadata fields </span> </a> <nav class=md-nav aria-label="Metadata fields"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-metadata-fields class=md-nav__link> <span class=md-ellipsis> General metadata fields </span> </a> </li> <li class=md-nav__item> <a href=#assay-metadata-fields class=md-nav__link> <span class=md-ellipsis> Assay metadata fields </span> </a> </li> <li class=md-nav__item> <a href=#project-metadata-fields class=md-nav__link> <span class=md-ellipsis> Project metadata fields </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#more-info class=md-nav__link> <span class=md-ellipsis> More info </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-naming-conventions class=md-nav__link> <span class=md-ellipsis> 3. Naming conventions </span> </a> <nav class=md-nav aria-label="3. Naming conventions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-tips class=md-nav__link> <span class=md-ellipsis> General tips </span> </a> </li> <li class=md-nav__item> <a href=#suggestions-for-ngs-data class=md-nav__link> <span class=md-ellipsis> Suggestions for NGS data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#4-create-a-catalog-of-your-assays-folder class=md-nav__link> <span class=md-ellipsis> 4. Create a catalog of your assays folder </span> </a> </li> <li class=md-nav__item> <a href=#5-version-control-of-your-data-analysis-using-git-and-github class=md-nav__link> <span class=md-ellipsis> 5. Version control of your data analysis using Git and Github </span> </a> <nav class=md-nav aria-label="5. Version control of your data analysis using Git and Github"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#creating-a-git-repo-online-and-copying-your-project-folder class=md-nav__link> <span class=md-ellipsis> Creating a git repo online and copying your project folder </span> </a> </li> <li class=md-nav__item> <a href=#github-pages class=md-nav__link> <span class=md-ellipsis> GitHub Pages </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#6-archive-github-repositories-on-zenodo class=md-nav__link> <span class=md-ellipsis> 6. Archive GitHub repositories on Zenodo </span> </a> <nav class=md-nav aria-label="6. Archive GitHub repositories on Zenodo"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#zenodo class=md-nav__link> <span class=md-ellipsis> Zenodo </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#wrap-up class=md-nav__link> <span class=md-ellipsis> Wrap up </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=keywords.html class=md-nav__link> <span class=md-ellipsis> Keyword index </span> </a> </li> <li class=md-nav__item> <a href=contributors.html class=md-nav__link> <span class=md-ellipsis> Contributors </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-organize-and-structure-your-ngs-data-and-data-analysis class=md-nav__link> <span class=md-ellipsis> 1. Organize and structure your NGS data and data analysis </span> </a> <nav class=md-nav aria-label="1. Organize and structure your NGS data and data analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#assay-folder class=md-nav__link> <span class=md-ellipsis> Assay folder </span> </a> </li> <li class=md-nav__item> <a href=#project-folder class=md-nav__link> <span class=md-ellipsis> Project folder </span> </a> </li> <li class=md-nav__item> <a href=#template-engine class=md-nav__link> <span class=md-ellipsis> Template engine </span> </a> <nav class=md-nav aria-label="Template engine"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#quick-tutorial-on-cookiecutter class=md-nav__link> <span class=md-ellipsis> Quick tutorial on cookiecutter </span> </a> <nav class=md-nav aria-label="Quick tutorial on cookiecutter"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-create-a-folder-template class=md-nav__link> <span class=md-ellipsis> Step 1: Create a Folder Template </span> </a> </li> <li class=md-nav__item> <a href=#step-2-create-cookiecutterjson class=md-nav__link> <span class=md-ellipsis> Step 2: Create cookiecutter.json </span> </a> </li> <li class=md-nav__item> <a href=#step-3-use-cookiecutter class=md-nav__link> <span class=md-ellipsis> Step 3: Use Cookiecutter </span> </a> </li> <li class=md-nav__item> <a href=#step-4-explore-the-generated-project class=md-nav__link> <span class=md-ellipsis> Step 4: Explore the Generated Project </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#2-metadata-and-naming-conventions class=md-nav__link> <span class=md-ellipsis> 2. Metadata and naming conventions </span> </a> <nav class=md-nav aria-label="2. Metadata and naming conventions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#readmemd-file class=md-nav__link> <span class=md-ellipsis> README.md file </span> </a> </li> <li class=md-nav__item> <a href=#metadatayml class=md-nav__link> <span class=md-ellipsis> metadata.yml </span> </a> </li> <li class=md-nav__item> <a href=#metadata-fields class=md-nav__link> <span class=md-ellipsis> Metadata fields </span> </a> <nav class=md-nav aria-label="Metadata fields"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-metadata-fields class=md-nav__link> <span class=md-ellipsis> General metadata fields </span> </a> </li> <li class=md-nav__item> <a href=#assay-metadata-fields class=md-nav__link> <span class=md-ellipsis> Assay metadata fields </span> </a> </li> <li class=md-nav__item> <a href=#project-metadata-fields class=md-nav__link> <span class=md-ellipsis> Project metadata fields </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#more-info class=md-nav__link> <span class=md-ellipsis> More info </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-naming-conventions class=md-nav__link> <span class=md-ellipsis> 3. Naming conventions </span> </a> <nav class=md-nav aria-label="3. Naming conventions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-tips class=md-nav__link> <span class=md-ellipsis> General tips </span> </a> </li> <li class=md-nav__item> <a href=#suggestions-for-ngs-data class=md-nav__link> <span class=md-ellipsis> Suggestions for NGS data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#4-create-a-catalog-of-your-assays-folder class=md-nav__link> <span class=md-ellipsis> 4. Create a catalog of your assays folder </span> </a> </li> <li class=md-nav__item> <a href=#5-version-control-of-your-data-analysis-using-git-and-github class=md-nav__link> <span class=md-ellipsis> 5. Version control of your data analysis using Git and Github </span> </a> <nav class=md-nav aria-label="5. Version control of your data analysis using Git and Github"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#creating-a-git-repo-online-and-copying-your-project-folder class=md-nav__link> <span class=md-ellipsis> Creating a git repo online and copying your project folder </span> </a> </li> <li class=md-nav__item> <a href=#github-pages class=md-nav__link> <span class=md-ellipsis> GitHub Pages </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#6-archive-github-repositories-on-zenodo class=md-nav__link> <span class=md-ellipsis> 6. Archive GitHub repositories on Zenodo </span> </a> <nav class=md-nav aria-label="6. Archive GitHub repositories on Zenodo"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#zenodo class=md-nav__link> <span class=md-ellipsis> Zenodo </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#wrap-up class=md-nav__link> <span class=md-ellipsis> Wrap up </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <nav class=md-tags> <a href=keywords.html#practical-lessons class=md-tag>Practical lessons</a> <a href=keywords.html#exercises class=md-tag>Exercises</a> <a href=keywords.html#dtu-workshop class=md-tag>DTU workshop</a> </nav> <div><h1 id=dtu-workshop-october-2023>DTU workshop October 2023<a class=headerlink href=#dtu-workshop-october-2023 title="Permanent link">¶</a></h1> <p><strong>Last updated:</strong> <em><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 30, 2023</span></em></p> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>🕰 <strong>Time Estimation:</strong> 75 minutes </p> <p>💬 <strong>Learning Objectives:</strong> </p> <ol> <li>Organize and structure your data and data analysis with cookiecutter templates</li> <li>Establish metadata fields and collect metadata when creating a cookiecutter folder</li> <li>Establish naming conventions for your data</li> <li>Make a catalog of your data</li> <li>Create GitHub repositories of your data analysis and display them as GitHub Pages</li> <li>Archive GitHub repositories on Zenodo</li> </ol> </div> <p>This is a practical version of the full RDM on NGS data workshop. The main key points of the exercises shown here is to help you organize and structure your NGS datasets and your data analyses. We will see how to keep track of your experiments metadata and how to safely version control and archive your data analyses using GitHub repositories and Zenodo. We hope that through these practical exercises and step-by-step guidance, you'll gain valuable skills in efficiently managing and sharing your research data, enhancing the reproducibility and impact of your work.</p> <div class="admonition warning"> <p class=admonition-title>Requirements</p> <p>In order to follow this tutorial you will need:</p> <ul> <li>A <a href=https://github.com/ >GitHub account</a> and git installed in your computer.</li> <li>A <a href=https://zenodo.org/ >Zenodo account</a></li> <li><a href=https://www.python.org/ >Python</a> and <a href=https://pip.pypa.io/en/stable/installation/ >pip</a> installed</li> <li><a href=https://www.mkdocs.org/ >MkDocs</a> and <a href=https://squidfunk.github.io/mkdocs-material/ >MkDocs material</a> theme (can be installed through pip)</li> <li><a href=https://cookiecutter.readthedocs.io/en/stable/ >Cookicutter</a> (can be installed through pip)</li> </ul> <p>In addition, you should install the following MkDocs extensions</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>pip<span class=w> </span>install<span class=w> </span>mkdocs
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>pip<span class=w> </span>install<span class=w> </span>mkdocs-material
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>pip<span class=w> </span>install<span class=w> </span>mkdocs-minify-plugin
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a>pip<span class=w> </span>install<span class=w> </span>mkdocs-git-revision-date-localized-plugin<span class=w> </span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>pip<span class=w> </span>install<span class=w> </span>mkdocs-jupyter
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>pip<span class=w> </span>install<span class=w> </span>mkdocs-table-reader-plugin
</code></pre></div> </div> <h2 id=1-organize-and-structure-your-ngs-data-and-data-analysis>1. Organize and structure your NGS data and data analysis<a class=headerlink href=#1-organize-and-structure-your-ngs-data-and-data-analysis title="Permanent link">¶</a></h2> <p>Applying a consistent file structure and naming conventions to your files will help you to efficiently manage your data. We will divide your NGS data and data analyses into two different types of folders:</p> <ol> <li><strong>Assay folders</strong>: These folders contain the <strong>raw and processed NGS datasets</strong>, as well as the <strong>pipeline/workflow</strong> used to generate the processed data, provenance of the raw data and quality control reports of the data. This data should be <strong>locked and read-only</strong> to prevent unwanted modifications.</li> <li><strong>Project folders</strong>: These folders contain <strong>all the necessary files for a specific research project</strong>. A project may use several assays or results from other projects. The assay data should not be copied or duplicated, but linked from the original source.</li> </ol> <p>Projects and Assays are separated from each other because a project may use one or more assays to answer a scientific question, and assays may be reused several times in different projects. This could be, for example, all the data analysis related to a publication (a RNAseq and a ChIPseq experiment), or a comparison between a previous ATACseq experiment (which was used for a older project) with a new laboratory protocol.</p> <p>You could also create <strong>Genomic resources folders</strong> such things such as genome references (fasta files) and annotations (gtf files) for different species, as well as indexes for different alignment algorithms. If you want to know more, feel free to check the relevant <a href=06_file_structure.html#genomic-resources-folder>full lesson</a></p> <p>This will help you to keep your data tidied up, specially if you are working on a big lab where assays may be used for different purposes and different people!</p> <h3 id=assay-folder>Assay folder<a class=headerlink href=#assay-folder title="Permanent link">¶</a></h3> <p>For each NGS experiment there should be an <code>Assay</code> folder that will contain all experimental datasets, that is, an <code>Assay</code> (raw files and pipeline processed files). Raw files should not be modified at all, but you should probably lock modifications to the final results once you are done with preprocessing the data. This will help you prevent unwanted modifications to the data. Each <code>Assay</code> subfolder should be named in a way that it is unique, easily readable, distinguishable and understood at a glance. For example, you could name an NGS assay using an acronym for the type of NGS assay (RNAseq, ChIPseq, ATACseq), a keyword that represents a unique descriptive element of that assay, and the date. Like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>&lt;Assay-ID&gt;_&lt;keyword&gt;_YYYYMMDD
</code></pre></div> <p>For example <code>CHIP_Oct4_20230101</code> is a ChIPseq assay made on 1<sup>st</sup> January 2023 with the keyword Oct4, so it is easily identifiable by eye. Next, let's take a look at a possible folder structure and what kind of files you can find there.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>CHIP_Oct4_20230101/
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>├──<span class=w> </span>README.md
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>├──<span class=w> </span>metadata.yml
<a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a>├──<span class=w> </span>pipeline.md
<a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a>├──<span class=w> </span>processed
<a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>└──<span class=w> </span>raw
<a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a><span class=w>   </span>├──<span class=w> </span>.fastq.gz
<a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a><span class=w>   </span>└──<span class=w> </span>samplesheet.csv
</code></pre></div> <ul> <li><strong>README.md</strong>: Long description of the assay in markdown format. It should contain provenance of the raw NGS data (samples, laboratory protocols used, aim of the assay, etc)</li> <li><strong>metadata.yml</strong>: metadata file for the assay describing different keys and important information regarding that assay (<a href=07_metadata.html>see this lesson</a>).</li> <li><strong>pipeline.md</strong>: description of the pipeline used to process raw data, as well as the commands used to run the pipeline.</li> <li><strong>processed</strong>: folder with results of the preprocessing pipeline. Contents depend on the pipeline used.</li> <li><strong>raw</strong>: folder with the raw data.<ul> <li><em>.fastq.gz</em>:In the case of NGS assays, there should be fastq files.</li> <li><em>samplesheet.csv</em>: file that contains metadata information for the samples. This file is used to run the nf-core pipelines. You can also add extra columns with info regarding the experimental variables and batches so it can be used for downstream analysis as well.</li> </ul> </li> </ul> <h3 id=project-folder>Project folder<a class=headerlink href=#project-folder title="Permanent link">¶</a></h3> <p>On the other hand, we have the other type of folder called <code>Projects</code>. In this folder you will save a <strong>subfolder for each project</strong> that you (or your lab) works on. Each <code>Project</code> subfolder will contain project information and all the data analysis notebooks and scripts used in that project.</p> <p>As like for an Assay folder, the Project folder should be named in a way that it is unique, easily readable, distinguishable and understood at a glance. For example, you could name it after the main author initials, a keyword that represents a unique descriptive element of that assay, and the date:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>&lt;author_initials&gt;_&lt;keyword&gt;_YYYYMMDD
</code></pre></div> <p>For example, <code>JARH_Oct4_20230101</code>, is a project about the gene Oct4 owned by Jose Alejandro Romero Herrera, created on the 1<sup>st</sup> of January of 2023.</p> <p>Next, let's take a look at a possible folder structure and what kind of files you can find there.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>&lt;author_initials&gt;_&lt;keyword&gt;_YYYYMMDD
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>├──<span class=w> </span>data
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>│<span class=w>  </span>└──<span class=w> </span>&lt;Assay-ID&gt;_&lt;keyword&gt;_YYYYMMDD/
<a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a>├──<span class=w> </span>documents
<a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>│<span class=w>  </span>└──<span class=w> </span>Non-sensitive_NGS_research_project_template.docx
<a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a>├──<span class=w> </span>notebooks
<a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis.rmd
<a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a>├──<span class=w> </span>README.md
<a id=__codelineno-4-9 name=__codelineno-4-9 href=#__codelineno-4-9></a>├──<span class=w> </span>reports
<a id=__codelineno-4-10 name=__codelineno-4-10 href=#__codelineno-4-10></a>│<span class=w>  </span>├──<span class=w> </span>figures
<a id=__codelineno-4-11 name=__codelineno-4-11 href=#__codelineno-4-11></a>│<span class=w>  </span>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis/
<a id=__codelineno-4-12 name=__codelineno-4-12 href=#__codelineno-4-12></a>│<span class=w>  </span>│<span class=w>   </span>└──<span class=w> </span>heatmap_sampleCor_20230102.png
<a id=__codelineno-4-13 name=__codelineno-4-13 href=#__codelineno-4-13></a>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis.html
<a id=__codelineno-4-14 name=__codelineno-4-14 href=#__codelineno-4-14></a>├──<span class=w> </span>requirements.txt
<a id=__codelineno-4-15 name=__codelineno-4-15 href=#__codelineno-4-15></a>├──<span class=w> </span>results
<a id=__codelineno-4-16 name=__codelineno-4-16 href=#__codelineno-4-16></a>│<span class=w>  </span>└──<span class=w> </span>01_data_analysis/
<a id=__codelineno-4-17 name=__codelineno-4-17 href=#__codelineno-4-17></a>│<span class=w>      </span>└──<span class=w> </span>DEA_treat-control_LFC1_p01.tsv
<a id=__codelineno-4-18 name=__codelineno-4-18 href=#__codelineno-4-18></a>├──<span class=w> </span>scripts
<a id=__codelineno-4-19 name=__codelineno-4-19 href=#__codelineno-4-19></a>└──<span class=w> </span>metadata.yml
</code></pre></div> <ul> <li><strong>data</strong>: folder that contains symlinks or shortcuts to where the data is, avoiding copying and modification of original files.</li> <li><strong>documents</strong>: folder containing word documents, slides or pdfs related to the project, such as explanations of the data or project, papers, etc. It also contains your <a href=05_DMP.html>Data Management Plan</a>.<ul> <li><em>Non-sensitive_NGS_research_project_template.docx</em>. This is a pre-filled Data Management Plan based on the Horizon Europe guidelines.</li> </ul> </li> <li><strong>notebooks</strong>: folder containing Jupyter, R markdown or Quarto notebooks with the actual data analysis.</li> <li><strong>README.md</strong>: detailed description of the project in markdown format.</li> <li><strong>reports</strong>: notebooks rendered as html/docx/pdf versions, ideal for sharing with colleagues and also as a formal report of the data analysis procedure.<ul> <li><em>figures</em>: figures produced upon rendering notebooks. The figures will be saved under a subfolder named after the notebook that created them. This is for provenance purposes so we know which notebook created which figures.</li> </ul> </li> <li><strong>requirements.txt</strong>: file explaining what software and libraries/packages and their versions are necessary to reproduce the code.</li> <li><strong>results</strong>: results from the data analysis, such as tables with differentially expressed genes, enrichment results, etc.</li> <li><strong>scripts</strong>: folder containing helper scripts needed to run data analysis or reproduce the work of the folder</li> <li><strong>description.yml</strong>: short description of the project.</li> <li><strong>metadata.yml</strong>: metadata file for the assay describing different keys (<a href=07_metadata.html>see this lesson</a>).</li> </ul> <h3 id=template-engine>Template engine<a class=headerlink href=#template-engine title="Permanent link">¶</a></h3> <p>It is very easy to create a folder template using <a href=https://github.com/cookiecutter/cookiecutter>cookiecutter</a>. Cookiecutter is a command-line utility that creates projects from cookiecutters (that is, a template), e.g. creating a Python package project from a Python package project template. Here you can find an example of a cookiecutter folder template directed to <a href=https://github.com/brickmanlab/ngs-template>NGS data</a>, where we have applied the structures explained in the previous sections. You are very welcome to adapt it or modify it to your needs!</p> <h4 id=quick-tutorial-on-cookiecutter>Quick tutorial on cookiecutter<a class=headerlink href=#quick-tutorial-on-cookiecutter title="Permanent link">¶</a></h4> <p>Creating a Cookiecutter template from scratch involves defining a folder structure, creating a <code>cookiecutter.json</code> file, and specifying the placeholders (keywords) that will be replaced during project generation. Let's walk through the process step by step:</p> <h5 id=step-1-create-a-folder-template>Step 1: Create a Folder Template<a class=headerlink href=#step-1-create-a-folder-template title="Permanent link">¶</a></h5> <p>Start by creating a folder with the structure you want for your template. For example, let's create a simple Python project template:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>my_template/
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>|-- {{cookiecutter.project_name}}
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>|   |-- main.py
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a>|-- tests
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a>|   |-- test_{{cookiecutter.project_name}}.py
<a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a>|-- README.md
</code></pre></div> <p>In this example, <code>{{cookiecutter.project_name}}</code> is a placeholder that will be replaced with the actual project name when the template is used.</p> <h5 id=step-2-create-cookiecutterjson>Step 2: Create <code>cookiecutter.json</code><a class=headerlink href=#step-2-create-cookiecutterjson title="Permanent link">¶</a></h5> <p>In the root of your template folder, create a file named <code>cookiecutter.json</code>. This file will define the variables (keywords) that users will be prompted to fill in. For our Python project template, it might look like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=p>{</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=w>  </span><span class=nt>"project_name"</span><span class=p>:</span><span class=w> </span><span class=s2>"MyProject"</span><span class=p>,</span>
<a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=w>  </span><span class=nt>"author_name"</span><span class=p>:</span><span class=w> </span><span class=s2>"Your Name"</span><span class=p>,</span>
<a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=w>  </span><span class=nt>"description"</span><span class=p>:</span><span class=w> </span><span class=s2>"A short description of your project"</span>
<a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=p>}</span>
</code></pre></div> <p>These are the questions users will be asked when generating a project based on your template. The values provided here will be used to replace the corresponding placeholders in the template files.</p> <p>In addition to replacing placeholders in file and directory names, Cookiecutter can also automatically fill in information within the contents of text files. This can be useful for providing default configurations or templates for code files. Let's extend our previous example to include a placeholder inside a text file:</p> <p>First, modify the <code>my_template/main.py</code> file to include a placeholder inside its contents:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># main.py</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=k>def</span> <span class=nf>hello</span><span class=p>():</span>
<a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a>    <span class=nb>print</span><span class=p>(</span><span class=s2>"Hello, {{cookiecutter.project_name}}!"</span><span class=p>)</span>
</code></pre></div> <p>Now, the <code>{{cookiecutter.project_name}}</code> placeholder is inside the <code>main.py</code> file. When you run Cookiecutter, it will automatically replace the placeholders not only in file and directory names but also within the contents of text files. After running Cookiecutter, your generated <code>main.py</code> file might look like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># main.py</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=k>def</span> <span class=nf>hello</span><span class=p>():</span>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a>    <span class=nb>print</span><span class=p>(</span><span class=s2>"Hello, MyProject!"</span><span class=p>)</span>  <span class=c1># Assuming "MyProject" was entered as the project_name</span>
</code></pre></div> <h5 id=step-3-use-cookiecutter>Step 3: Use Cookiecutter<a class=headerlink href=#step-3-use-cookiecutter title="Permanent link">¶</a></h5> <p>Now that your template is set up, you can use Cookiecutter to generate a project based on it. Open a terminal and run:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a>cookiecutter<span class=w> </span>path/to/your/template
</code></pre></div> <p>Cookiecutter will prompt you to fill in the values for <code>project_name</code>, <code>author_name</code>, and <code>description</code>. After you provide these values, Cookiecutter will replace the placeholders in your template files with the entered values.</p> <h5 id=step-4-explore-the-generated-project>Step 4: Explore the Generated Project<a class=headerlink href=#step-4-explore-the-generated-project title="Permanent link">¶</a></h5> <p>Once the generation process is complete, navigate to the directory where Cookiecutter created the new project. You will see a project structure with the placeholders replaced by the values you provided.</p> <div class="admonition question"> <p class=admonition-title>Exercise 1: Create your own template</p> <p>Using cookiecutter, create your own templates for your folders. You do not need to copy exactly our suggestions, adjust your template to your own needs!</p> <p>We have prepared already two simple cookiecutter templates in GitHub repositories.</p> <p><strong>Assay</strong></p> <ol> <li>First, fork our <a href=https://github.com/hds-sandbox/assay-template>Assay folder template</a> from the GitHub page into your own account/organization. <img alt=fork_repo_example src=images/fork_repo.png></li> <li>Then, use <code>git clone &lt;your URL to the template&gt;</code> to put it in your computer.</li> <li>Modify the contents of the repository so that it matches the <strong>Assay</strong> example above. You are welcome to do changes as you please!</li> <li>Modify the <code>cookiecutter.json</code> file so that it will include the <strong>Assay</strong> name template</li> <li>Git add, commit and push your changes</li> <li>Test your folder by using <code>cookiecutter &lt;URL to your GitHub repository for "assay-template&gt;</code></li> </ol> <p><strong>Project</strong></p> <ol> <li>First, fork our <a href=https://github.com/hds-sandbox/project-template>Project folder template</a> from the GitHub page into your own account/organization. <img alt=fork_repo_example src=images/fork_repo_project.png></li> <li>Then, use <code>git clone &lt;your URL to the template&gt;</code> to put it in your computer.</li> <li>Modify the contents of the repository so that it matches the <strong>Project</strong> example above. You are welcome to do changes as you please!</li> <li>Modify the <code>cookiecutter.json</code> file so that it will include the <strong>Project</strong> name template</li> <li>Git add, commit and push your changes</li> <li>Test your folder by using <code>cookiecutter &lt;URL to your GitHub repository for "project-template&gt;</code></li> </ol> </div> <h2 id=2-metadata-and-naming-conventions>2. Metadata and naming conventions<a class=headerlink href=#2-metadata-and-naming-conventions title="Permanent link">¶</a></h2> <p>Metadata is the behind-the-scenes information that makes sense of data and gives context and structure. For NGS data, metadata includes information such as when and where the data was collected, what it represents, and how it was processed. Let's check what kind of relevant metadata is available for NGS data and how to capture it in your Assay or Project folders. Both of these folders contain a metadata.yml file and a README.md file. In this section, we will check what kind of information you should collect in each of these files.</p> <div class="admonition warning"> <p class=admonition-title>Metadata and controlled vocabularies</p> <p>In order for metadata to be most useful, you should try to use controlled vocabularies for all your fields. For example, tissue could be described with the <a href=https://www.ebi.ac.uk/ols/ontologies/uberon>UBERON ontologies</a>, species using the <a href=https://www.ncbi.nlm.nih.gov/taxonomy>NCBI taxonomy</a>, diseases using the <a href=https://mondo.monarchinitiative.org/ >Mondo database</a>, etc. Unfortunately, implementing a systematic way of using these vocabularies is rather complex and outside the scope of this workshop, but you are very welcome to try to implement them on your own!</p> </div> <h3 id=readmemd-file>README.md file<a class=headerlink href=#readmemd-file title="Permanent link">¶</a></h3> <p>The README.md file is a <a href=https://www.markdownguide.org/ >markdown file</a> that allows you to write a long description of the data placed in a folder. Since it is a markdown file, you are able to write in rich text format (bold, italic, include links, etc) what is inside the folder, why it was created/collected, how and when. If it is an <code>Assay</code> folder, you could include the laboratory protocol used to generate the samples, images explaining the experiment design, a summary of the results of the experiment and any sort of comments that would help to understand the context of the experiment. On the other hand, a 'Project' README file may contain a description of the project, what are its aims, why is it important, what 'Assays' is it using, how to interpret the code notebooks, a summary of the results and, again, any sort of comments that would help to understand the project.</p> <p>Here is an example of a README file for a `Project`` folder:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a># NGS Analysis Project: Exploring Gene Expression in Human Tissues
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>
<a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a>## Aims
<a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a>
<a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a>This project aims to investigate gene expression patterns across various human tissues using Next Generation Sequencing (NGS) data. By analyzing the transcriptomes of different tissues, we seek to uncover tissue-specific gene expression profiles and identify potential markers associated with specific biological functions or diseases.
<a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a>
<a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a>## Why It's Important
<a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a>
<a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a>Understanding tissue-specific gene expression is crucial for deciphering the molecular basis of health and disease. Identifying genes that are uniquely expressed in certain tissues can provide insights into tissue function, development, and potential therapeutic targets. This project contributes to our broader understanding of human biology and has implications for personalized medicine and disease research.
<a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a>
<a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a>## Datasets
<a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a>
<a id=__codelineno-10-13 name=__codelineno-10-13 href=#__codelineno-10-13></a>We have used internal datasets with IDs: RNA_humanSkin_20201030, RNA_humanBrain_20210102, RNA_humanLung_20220304.
<a id=__codelineno-10-14 name=__codelineno-10-14 href=#__codelineno-10-14></a>
<a id=__codelineno-10-15 name=__codelineno-10-15 href=#__codelineno-10-15></a>In addition, we utilized publicly available NGS datasets from the GTEx (Genotype-Tissue Expression) project, which provides comprehensive RNA-seq data across multiple human tissues. These datasets offer a wealth of information on gene expression levels and isoform variations across diverse tissues, making them ideal for our analysis.
<a id=__codelineno-10-16 name=__codelineno-10-16 href=#__codelineno-10-16></a>
<a id=__codelineno-10-17 name=__codelineno-10-17 href=#__codelineno-10-17></a>## Summary of Results
<a id=__codelineno-10-18 name=__codelineno-10-18 href=#__codelineno-10-18></a>
<a id=__codelineno-10-19 name=__codelineno-10-19 href=#__codelineno-10-19></a>Our analysis revealed distinct gene expression patterns among different human tissues. We identified tissue-specific genes enriched in brain tissues, highlighting their potential roles in neurodevelopment and function. Additionally, we found a set of genes that exhibit consistent expression across a range of tissues, suggesting their fundamental importance in basic cellular processes.
<a id=__codelineno-10-20 name=__codelineno-10-20 href=#__codelineno-10-20></a>
<a id=__codelineno-10-21 name=__codelineno-10-21 href=#__codelineno-10-21></a>Furthermore, our differential expression analysis unveiled significant changes in gene expression between healthy and diseased tissues, shedding light on potential molecular factors underlying various diseases. Overall, this project underscores the power of NGS data in unraveling intricate gene expression networks and their implications for human health.
<a id=__codelineno-10-22 name=__codelineno-10-22 href=#__codelineno-10-22></a>
<a id=__codelineno-10-23 name=__codelineno-10-23 href=#__codelineno-10-23></a>---
<a id=__codelineno-10-24 name=__codelineno-10-24 href=#__codelineno-10-24></a>
<a id=__codelineno-10-25 name=__codelineno-10-25 href=#__codelineno-10-25></a>For more details, refer to our [Jupyter Notebook](link-to-jupyter-notebook.ipynb) for the complete analysis pipeline and code.
</code></pre></div> <h3 id=metadatayml>metadata.yml<a class=headerlink href=#metadatayml title="Permanent link">¶</a></h3> <p>The metadata file is a <a href=https://fileinfo.com/extension/yml>yml file</a>, which is a text document that contains data formatted using a human-readable data format for data serialization.</p> <p><img alt="yaml file example" src=images/yml_example.png></p> <h3 id=metadata-fields>Metadata fields<a class=headerlink href=#metadata-fields title="Permanent link">¶</a></h3> <p>There is a ton of information you can collect regarding an NGS assay or a project. Some information fields are very general, such as author or date, while others are specific to the Assay or Project folder. Below, we will take a look at minimal information you should collect in each of the folders.</p> <h4 id=general-metadata-fields>General metadata fields<a class=headerlink href=#general-metadata-fields title="Permanent link">¶</a></h4> <p>Here you can find a list of suggestions for general metadata fields that can be used for both assays and project folders:</p> <ul> <li><strong>Title</strong>: A brief yet informative name for the dataset.</li> <li><strong>Author(s)</strong>: The individual(s) or organization responsible for creating the dataset. You can use your <a href=https://orcid.org/ >ORCID</a></li> <li><strong>Date Created</strong>: The date when the dataset was originally generated or compiled. Use YYYY-MM-DD format!</li> <li><strong>Description</strong>: A short narrative explaining the content, purpose, and context.</li> <li><strong>Keywords</strong>: A set of descriptive terms or phrases that capture the folder's main topics and attributes.</li> <li><strong>Version</strong>: The version number or identifier for the folder, useful for tracking changes.</li> <li><strong>License</strong>: The type of license or terms of use associated with the dataset/project.</li> </ul> <h4 id=assay-metadata-fields>Assay metadata fields<a class=headerlink href=#assay-metadata-fields title="Permanent link">¶</a></h4> <p>Here you will find a table with possible metadata fields that you can use to annotate and track your <code>Assay</code> folders:</p> <table> <thead> <tr> <th style="text-align: left;">Metadata field</th> <th style="text-align: left;">Definition</th> <th style="text-align: left;">Format</th> <th style="text-align: left;">Ontology</th> <th style="text-align: left;">Example</th> </tr> </thead> <tbody> <tr> <td style="text-align: left;">assay_ID</td> <td style="text-align: left;">Identifier for the assay that is at least unique within the project</td> <td style="text-align: left;">&lt;Assay-ID&gt;_&lt;keyword&gt;_YYYYMMDD</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">CHIP_Oct4_20200101</td> </tr> <tr> <td style="text-align: left;">assay_type</td> <td style="text-align: left;">The type of experiment performed, eg ATAC-seq or seqFISH</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">ontology field- e.g. EFO or OBI</td> <td style="text-align: left;">ChIPseq</td> </tr> <tr> <td style="text-align: left;">assay_subtype</td> <td style="text-align: left;">More specific type or assay like bulk nascent RNAseq or single cell ATACseq</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">ontology field- e.g. EFO or OBI</td> <td style="text-align: left;">bulk ChIPseq</td> </tr> <tr> <td style="text-align: left;">owner</td> <td style="text-align: left;">Owner of the assay (who made the experiment?).</td> <td style="text-align: left;">&lt;First Name&gt; &lt;Last Name&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">Jose Romero</td> </tr> <tr> <td style="text-align: left;">platform</td> <td style="text-align: left;">The type of instrument used to perform the assay, eg Illumina HiSeq 4000 or Fluidigm C1 microfluidics platform</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">ontology field- e.g. EFO or OBI</td> <td style="text-align: left;">Illumina</td> </tr> <tr> <td style="text-align: left;">extraction_method</td> <td style="text-align: left;">Technique used to extract the nucleic acid from the cell</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">ontology field- e.g. EFO or OBI</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">library_method</td> <td style="text-align: left;">Technique used to amplify a cDNA library</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">ontology field- e.g. EFO or OBI</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">external_accessions</td> <td style="text-align: left;">Accession numbers from external resources to which assay or protocol information was submitted</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">eg protocols.io, AE, GEO accession number, etc</td> <td style="text-align: left;">GSEXXXXX</td> </tr> <tr> <td style="text-align: left;">keyword</td> <td style="text-align: left;">Keyword for easy identification</td> <td style="text-align: left;">wordWord</td> <td style="text-align: left;">camelCase</td> <td style="text-align: left;">Oct4ChIP</td> </tr> <tr> <td style="text-align: left;">date</td> <td style="text-align: left;">Date of assay creation</td> <td style="text-align: left;">YYYYMMDD</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">20200101</td> </tr> <tr> <td style="text-align: left;">nsamples</td> <td style="text-align: left;">Number of samples analyzed in this assay</td> <td style="text-align: left;">&lt;integer&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">9</td> </tr> <tr> <td style="text-align: left;">is_paired</td> <td style="text-align: left;">Paired fastq files or not</td> <td style="text-align: left;">&lt;single OR paired&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">single</td> </tr> <tr> <td style="text-align: left;">pipeline</td> <td style="text-align: left;">Pipeline used to process data and version</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nf-core/chipseq -r 1.0</td> </tr> <tr> <td style="text-align: left;">strandedness</td> <td style="text-align: left;">The strandedness of the cDNA library</td> <td style="text-align: left;">&lt;+ OR - OR *&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">*</td> </tr> <tr> <td style="text-align: left;">processed_by</td> <td style="text-align: left;">Who processed the data</td> <td style="text-align: left;">&lt;First Name&gt; &lt;Last Name&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">Sarah Lundregan</td> </tr> <tr> <td style="text-align: left;">organism</td> <td style="text-align: left;">Organism origin</td> <td style="text-align: left;">&lt;Genus species&gt;</td> <td style="text-align: left;">Taxonomy name</td> <td style="text-align: left;">Mus musculus</td> </tr> <tr> <td style="text-align: left;">origin</td> <td style="text-align: left;">Is internal or external (from a public resources) data</td> <td style="text-align: left;">&lt;internal OR external&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">internal</td> </tr> <tr> <td style="text-align: left;">path</td> <td style="text-align: left;">Path to files</td> <td style="text-align: left;">&lt;/path/to/file&gt;</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">short_desc</td> <td style="text-align: left;">Short description of the assay</td> <td style="text-align: left;">plain text</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">Oct4 ChIP after pERK activation</td> </tr> <tr> <td style="text-align: left;">ELN_ID</td> <td style="text-align: left;">ID of the experiment/assay in your Electronic Lab Notebook software, like labguru or benchling</td> <td style="text-align: left;">plain text</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> </tr> </tbody> </table> <h4 id=project-metadata-fields>Project metadata fields<a class=headerlink href=#project-metadata-fields title="Permanent link">¶</a></h4> <p>Here you will find a table with possible metadata fields that you can use to annotate and track your <code>Project</code> folders:</p> <table> <thead> <tr> <th style="text-align: left;">Metadata field</th> <th style="text-align: left;">Definition</th> <th style="text-align: left;">Format</th> <th style="text-align: right;">Ontology</th> <th style="text-align: left;">Example</th> </tr> </thead> <tbody> <tr> <td style="text-align: left;">project</td> <td style="text-align: left;">Project ID</td> <td style="text-align: left;">&lt;surname&gt;_et_al_2023</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">proks_et_al_2023</td> </tr> <tr> <td style="text-align: left;">author</td> <td style="text-align: left;">Owner of the project</td> <td style="text-align: left;">&lt;First name&gt; &lt;Surname&gt;</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">Martin Proks</td> </tr> <tr> <td style="text-align: left;">date</td> <td style="text-align: left;">Date of creation</td> <td style="text-align: left;">YYYYMMDD</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">20230101</td> </tr> <tr> <td style="text-align: left;">description</td> <td style="text-align: left;">Short description of the project</td> <td style="text-align: left;">Plain text</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">This is a project describing the effect of Oct4 perturbation after pERK activation</td> </tr> </tbody> </table> <h3 id=more-info>More info<a class=headerlink href=#more-info title="Permanent link">¶</a></h3> <p>The information provided in this lesson is not at all exhaustive. There might be many more fields and controlled vocabularies that could be useful for your NGS data. We recommend that you take a look at the following sources for more information!</p> <ul> <li><a href=https://faircookbook.elixir-europe.org/content/recipes/interoperability/transcriptomics-metadata.html#analysis-metadata>Transcriptomics metadata standards and fields</a></li> <li><a href=https://lamin.ai/docs/bionty>Bionty</a>: Biological ontologies for data scientists.</li> </ul> <div class="admonition question"> <p class=admonition-title>Exercise 2: modify the metadata.yml files in your cookiecutter templates</p> <p>We have seen some examples of metadata for NGS data. It is time now to customize your cookiecutter templates and modify the metadata.yml files so that they fit your needs! </p> <ol> <li>Think about what kind of metadata you would like to include.</li> <li>Modify the <code>cookiecutter.json</code> file so that when you create a new folder template, all the metadata is filled accordingly. <img alt=cookiecutter_json_example src=images/cookiecutter_json.png></li> <li>Modify the <code>metadata.yml</code> file so that it includes the metadata recorded by the <code>cookiecutter.json</code> file. <img alt=assay_metadata_example src=images/assay_metadata.png></li> <li>Modify the <code>README.md</code> file so that it includes the short description recorded by the <code>cookiecutter.json</code> file.</li> <li>Git add, commit and push the changes of your template.</li> <li>Test your folders by using the command <code>cookiecutter &lt;URL to your cookiecutter repository in GitHub&gt;</code></li> </ol> </div> <h2 id=3-naming-conventions>3. Naming conventions<a class=headerlink href=#3-naming-conventions title="Permanent link">¶</a></h2> <p>Using consistent naming conventions is important in scientific research as it helps with the organization and retrieval of data or results. By adopting standardized naming conventions, researchers ensure that files, experiments, or data sets are labeled in a clear, logical manner. This makes it easier to locate and compare similar types of data or results, even when dealing with large datasets or multiple experiments. For instance, in genomics, employing uniform naming conventions for files related to specific experiments or samples allows for swift identification and comparison of relevant data, streamlining the research process and contributing to the reproducibility of findings. This practice promotes efficiency, collaboration, and the integrity of scientific work.</p> <h3 id=general-tips>General tips<a class=headerlink href=#general-tips title="Permanent link">¶</a></h3> <p>Below you will find a small list of general tips to follow when you name a folder or a file:</p> <ul> <li>Use only alphanumeric characters to write a word: a to z and 0 to 9</li> <li>Avoid special characters: ~!@#$%^&amp;*()`[]{}"|</li> <li>Date format: use <code>YYYYMMDD</code> format. For example: 20230101.</li> <li>Authors: use initials. For example: JARH</li> <li><strong>Don't use of spaces</strong>! Computers get very confused when you need to point a path to a file and it contains spaces! Instead:<ul> <li>Separate field sections are separated by underscores <code>_</code>.</li> <li>Words in each section are written in <a href=https://en.wikipedia.org/wiki/Camel_case>camelCase</a>. It would look then like this: <code>field1_word1Word2.txt</code>. For example: <code>heatmap_sampleCor_20230101.png</code>. The first field indicates what this file is, i.e., a heatmap. Second field is what is being plotted, i.e, sample correlations; since the field contains two words, they are written in camelCase. The third field is the date of when the image was created.</li> </ul> </li> <li>Use as short fields as possible. You can try to use understandable abbreviations, like LFC for LogFoldChange, Cor for correlations, Dist for distances, etc.</li> <li>Avoid long names as much as you can, be concise!</li> <li>Avoid creating many sublevels of folders.</li> <li>Write down your naming convention pattern and document it in the README file</li> <li>When using a sequential numbering system, use leading zeros to make sure files sort in sequential order. Using <code>01</code> instead of just <code>1</code> if your sequence only goes up to <code>99</code>.</li> <li>Versions should be used as the last element, and use at least two digits with a leading 0 (e.g. v01, v02)</li> </ul> <h3 id=suggestions-for-ngs-data>Suggestions for NGS data<a class=headerlink href=#suggestions-for-ngs-data title="Permanent link">¶</a></h3> <p>More info on naming conventions for different types of files and analysis is in development.</p> <table> <thead> <tr> <th style="text-align: left;">name</th> <th style="text-align: left;">description</th> <th style="text-align: left;">naming_convention</th> <th style="text-align: left;">file format</th> <th style="text-align: left;">example</th> </tr> </thead> <tbody> <tr> <td style="text-align: left;">.fastq</td> <td style="text-align: left;">raw sequencing reads</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">sampleID_run_read1.fastq</td> </tr> <tr> <td style="text-align: left;">.fastqc</td> <td style="text-align: left;">quality control from fastqc</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">sampleID_run_read1.fastqc</td> </tr> <tr> <td style="text-align: left;">.bam</td> <td style="text-align: left;">aligned reads</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">sampleID_run_read1.bam</td> </tr> <tr> <td style="text-align: left;">GTF</td> <td style="text-align: left;">sequence annotation</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">one of https://www.gencodegenes.org/</td> </tr> <tr> <td style="text-align: left;">GFF</td> <td style="text-align: left;">sequence annotation</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">one of https://www.gencodegenes.org/</td> </tr> <tr> <td style="text-align: left;">.bed</td> <td style="text-align: left;">genome locations</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">.bigwig</td> <td style="text-align: left;">genome coverage</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> </tr> <tr> <td style="text-align: left;">.fasta</td> <td style="text-align: left;">sequence data (nucleotide/aminoacid)</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">one of https://www.gencodegenes.org/</td> </tr> <tr> <td style="text-align: left;">Multiqc report</td> <td style="text-align: left;">QC aggregated report</td> <td style="text-align: left;">&lt;assayID&gt;_YYYYMMDD.multiqc</td> <td style="text-align: left;">multiqc</td> <td style="text-align: left;">RNA_20200101.multiqc</td> </tr> <tr> <td style="text-align: left;">Count matrix</td> <td style="text-align: left;">final count matrix</td> <td style="text-align: left;">&lt;assayID&gt;_cm_aligner_YYYYMMDD.tsv</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">RNA_cm_salmon_20200101.tsv</td> </tr> <tr> <td style="text-align: left;">DEA</td> <td style="text-align: left;">differential expression analysis results</td> <td style="text-align: left;">DEA_&lt;condition1-condition2&gt;_LFC&lt;absolute_threshold&gt;_p&lt;pvalue decimals&gt;_YYYYMMDD.tsv</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">DEA_treat-untreat_LFC1_p01_20200101.tsv</td> </tr> <tr> <td style="text-align: left;">DBA</td> <td style="text-align: left;">differential binding analysis results</td> <td style="text-align: left;">DBA_&lt;condition1-condition2&gt;_LFC&lt;absolute_threshold&gt;_p&lt;pvalue decimals&gt;_YYYYMMDD.tsv</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">DBA_treat-untreat_LFC1_p01_20200101.tsv</td> </tr> <tr> <td style="text-align: left;">MAplot</td> <td style="text-align: left;">MA plot</td> <td style="text-align: left;">MAplot_&lt;condition1-condition2&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">MAplot_treat-untreat_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Heatmap plot</td> <td style="text-align: left;">Heatmap plot of anything</td> <td style="text-align: left;">heatmap_&lt;type&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">heatmap_sampleCor_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Volcano plot</td> <td style="text-align: left;">Volcano plot</td> <td style="text-align: left;">volcano_&lt;condition1-condition2&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">volcano_treat-untreat_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Venn diagram</td> <td style="text-align: left;">Venn diagram</td> <td style="text-align: left;">venn_&lt;type&gt;_YYYYMMDD.jpeg</td> <td style="text-align: left;">jpeg</td> <td style="text-align: left;">venn_consensus_20200101.jpeg</td> </tr> <tr> <td style="text-align: left;">Enrichment table</td> <td style="text-align: left;">Enrichment results</td> <td style="text-align: left;">nan</td> <td style="text-align: left;">tsv</td> <td style="text-align: left;">nan</td> </tr> </tbody> </table> <div class="admonition question"> <p class=admonition-title>Exercise 3: Create your own naming conventions</p> <p>Think about the most common types of files and folders you will be working on, such as visualizations, results tables, processed files, etc. Then come up with a logical and clear way of naming those files using the tips suggested above. Remember to avoid making long and complicated names!</p> </div> <h2 id=4-create-a-catalog-of-your-assays-folder>4. Create a catalog of your assays folder<a class=headerlink href=#4-create-a-catalog-of-your-assays-folder title="Permanent link">¶</a></h2> <p>The next step is to collect all the NGS datasets that you have created in the manner explained above. Since your folders all should contain the <code>metadata.yml</code> file in the same place with the same metadata, it should be very easy to iteratively go through all the folders and merge all the metadata.yml files into a one single table. This table can be then browsed easily with Microsoft Excel, for example. If you are interested in making a Shiny app or Python Panel tool to interactively browse the catalog, check out this <a href=08_database.html>lesson</a>.</p> <div class="admonition question"> <p class=admonition-title>Exercise 4: create a metadata.tsv catalog</p> <p>We will make a small script in R (or you can make one with python) that recursively goes through all the folders inside a input path (like your <code>Assays</code> folder), fetch all the <code>metadata.yml</code> files and merge them. Finally, it will write a tsv file as an output. </p> <ol> <li>Create a folder call <code>Assays</code></li> <li>Under that folder, make three new <code>Assay</code> folders from your cookiecutter template</li> <li>Run the script below with R (or create your own with python). Modify the <code>folder_path</code> variable so it matches the path tot the folder <code>Assays</code>. The table will be written under the same <code>folder_path</code>.</li> <li>Visualize your <code>Assays</code> table with Excel</li> </ol> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=nf>library</span><span class=p>(</span><span class=n>yaml</span><span class=p>)</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=nf>library</span><span class=p>(</span><span class=n>dplyr</span><span class=p>)</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=nf>library</span><span class=p>(</span><span class=n>lubridate</span><span class=p>)</span>
<a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a>
<a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a><span class=c1># Function to recursively fetch metadata.yml files</span>
<a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a><span class=n>get_metadata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>function</span><span class=p>(</span><span class=n>folder_path</span><span class=p>)</span><span class=w> </span><span class=p>{</span>
<a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a><span class=w>    </span><span class=n>file_list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list.files</span><span class=p>(</span><span class=n>path</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>folder_path</span><span class=p>,</span><span class=w> </span><span class=n>pattern</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"metadata\\.yml$"</span><span class=p>,</span><span class=w> </span><span class=n>recursive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>,</span><span class=w> </span><span class=n>full.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-11-8 name=__codelineno-11-8 href=#__codelineno-11-8></a><span class=w>    </span><span class=n>metadata_list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>lapply</span><span class=p>(</span><span class=n>file_list</span><span class=p>,</span><span class=w> </span><span class=n>yaml</span><span class=o>::</span><span class=n>yaml.load_file</span><span class=p>)</span>
<a id=__codelineno-11-9 name=__codelineno-11-9 href=#__codelineno-11-9></a><span class=w>    </span><span class=nf>return</span><span class=p>(</span><span class=n>metadata_list</span><span class=p>)</span>
<a id=__codelineno-11-10 name=__codelineno-11-10 href=#__codelineno-11-10></a><span class=w>    </span><span class=p>}</span>
<a id=__codelineno-11-11 name=__codelineno-11-11 href=#__codelineno-11-11></a>
<a id=__codelineno-11-12 name=__codelineno-11-12 href=#__codelineno-11-12></a><span class=c1># Specify the folder path</span>
<a id=__codelineno-11-13 name=__codelineno-11-13 href=#__codelineno-11-13></a><span class=n>folder_path</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=s>"/path/to/your/folder"</span>
<a id=__codelineno-11-14 name=__codelineno-11-14 href=#__codelineno-11-14></a>
<a id=__codelineno-11-15 name=__codelineno-11-15 href=#__codelineno-11-15></a><span class=c1># Fetch metadata from the specified folder</span>
<a id=__codelineno-11-16 name=__codelineno-11-16 href=#__codelineno-11-16></a><span class=n>metadata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>get_metadata</span><span class=p>(</span><span class=n>folder_path</span><span class=p>)</span>
<a id=__codelineno-11-17 name=__codelineno-11-17 href=#__codelineno-11-17></a>
<a id=__codelineno-11-18 name=__codelineno-11-18 href=#__codelineno-11-18></a><span class=c1># Convert metadata to a data frame</span>
<a id=__codelineno-11-19 name=__codelineno-11-19 href=#__codelineno-11-19></a><span class=n>metadata_df</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>data.frame</span><span class=p>(</span><span class=nf>matrix</span><span class=p>(</span><span class=nf>unlist</span><span class=p>(</span><span class=n>metadata</span><span class=p>),</span><span class=w> </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>length</span><span class=p>(</span><span class=n>metadata</span><span class=p>),</span><span class=w> </span><span class=n>byrow</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>))</span>
<a id=__codelineno-11-20 name=__codelineno-11-20 href=#__codelineno-11-20></a><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata_df</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>names</span><span class=p>(</span><span class=n>metadata</span><span class=p>[[</span><span class=m>1</span><span class=p>]])</span>
<a id=__codelineno-11-21 name=__codelineno-11-21 href=#__codelineno-11-21></a>
<a id=__codelineno-11-22 name=__codelineno-11-22 href=#__codelineno-11-22></a><span class=c1># Save the data frame as a TSV file</span>
<a id=__codelineno-11-23 name=__codelineno-11-23 href=#__codelineno-11-23></a><span class=n>output_file</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>paste0</span><span class=p>(</span><span class=s>"database_"</span><span class=p>,</span><span class=w> </span><span class=nf>format</span><span class=p>(</span><span class=nf>Sys.Date</span><span class=p>(),</span><span class=w> </span><span class=s>"%Y%m%d"</span><span class=p>),</span><span class=w> </span><span class=s>".tsv"</span><span class=p>)</span>
<a id=__codelineno-11-24 name=__codelineno-11-24 href=#__codelineno-11-24></a><span class=nf>write.table</span><span class=p>(</span><span class=n>metadata_df</span><span class=p>,</span><span class=w> </span><span class=n>file</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>output_file</span><span class=p>,</span><span class=w> </span><span class=n>sep</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"\t"</span><span class=p>,</span><span class=w> </span><span class=n>quote</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span><span class=n>row.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-11-25 name=__codelineno-11-25 href=#__codelineno-11-25></a>
<a id=__codelineno-11-26 name=__codelineno-11-26 href=#__codelineno-11-26></a><span class=c1># Print confirmation message</span>
<a id=__codelineno-11-27 name=__codelineno-11-27 href=#__codelineno-11-27></a><span class=nf>cat</span><span class=p>(</span><span class=s>"Database saved as"</span><span class=p>,</span><span class=w> </span><span class=n>output_file</span><span class=p>,</span><span class=w> </span><span class=s>"\n"</span><span class=p>)</span>
</code></pre></div> </div> <h2 id=5-version-control-of-your-data-analysis-using-git-and-github>5. Version control of your data analysis using Git and Github<a class=headerlink href=#5-version-control-of-your-data-analysis-using-git-and-github title="Permanent link">¶</a></h2> <p>Version control is a systematic approach to tracking changes made to a project over time. It provides a structured means of documenting alterations, allowing you to revisit and understand the evolution of your work. In research data management and data analytics, version control is very important and gives you a lot of advantages.</p> <p><a href=https://git-scm.com/about>Git</a> is a distributed version control system that enables developers and researchers to efficiently manage their project's history, collaborate seamlessly, and ensure data integrity. At its core, Git operates through the following principles and mechanisms: On the other hand, <a href=https://github.com/ >GitHub</a> is a web-based platform that enhances Git's capabilities by providing a collaborative and centralized hub for hosting Git repositories. It offers several key functionalities, such as tracking issues, security features to safeguard your repos, and GitHub Pages that allows you to create websites to showcase your projects.</p> <div class="admonition tip"> <p class=admonition-title>Create a GitHub organization for your lab or department</p> <p>GitHub allows users to create organizations and teams that will collaborate together or create repositories under the same umbrella organization. If you would like to create an educational organization in GitHub, you can do so for free! For example, you could create a GitHub account for your lab.</p> <p>In order to create a GitHub organization, follow these <a href=https://docs.github.com/en/organizations/collaborating-with-groups-in-organizations/creating-a-new-organization-from-scratch>instructions</a></p> <p>After you have created the GitHub organization, make sure that you create your repositories under the organization space and not your own user!</p> </div> <h3 id=creating-a-git-repo-online-and-copying-your-project-folder>Creating a git repo online and copying your project folder<a class=headerlink href=#creating-a-git-repo-online-and-copying-your-project-folder title="Permanent link">¶</a></h3> <p>Version controlling your data analysis folders, a.k.a. <code>Project</code> folder, is very easy once you have set up your cookiecutter templates. The simplest way of doing this is to first create a remote GitHub repository from the webpage (or from the Desktop app, if you are using it) with a proper project name. Then <code>git clone</code> that repository you just made into your <code>Projects</code> main folder. Then, use cookiecutter to create a project folder template and copy-paste the contents of the folder template to your cloned repo. If you wish, you could already git add, commit and push the first changes to the folders and continue from there on.</p> <div class="admonition tip"> <p class=admonition-title>Tips to write good commit messages</p> <p>If you would like to know more about Git commits and the best way to make clear git messages, check out <a href=https://www.conventionalcommits.org/en/v1.0.0/ >this post</a>!</p> </div> <h3 id=github-pages>GitHub Pages<a class=headerlink href=#github-pages title="Permanent link">¶</a></h3> <p>Once you have created your repository (and put it in GitHub), you have now the opportunity to add your data analysis reports that you created, in either Jupyter Notebooks, Rmarkdowns or html reports, in a <a href=https://pages.github.com/ >GitHub Page website</a>. Creating a GitHub page is very simple, and we really recommend that you follow the nice tutorial that GitHub as put for you. Nonetheless, we will see the main steps in the exercise below.</p> <p>There are many different ways to create your webpages. We recommend using Mkdocs and Mkdocs materials as a framework to create a nice webpage in a simple manner. The folder templates that we used as an example in the previous exercise already contain everything you need to start a webpage. Nonetheless, you will need to understand the basics of <a href=https://www.mkdocs.org/ >MkDocs</a> and <a href=https://squidfunk.github.io/mkdocs-material/ >MkDocs materials</a> to design a webpage to your liking. MkDocs is a static webpage generator that is very easy to use, while MkDocs materials is an extension of the tool that gives you many more options to customize your website. Check out their webpages to get started!</p> <div class="admonition question"> <p class=admonition-title>Exercise 5: make a project folder and publish a data analysis webpage</p> <ol> <li>Configure your main GitHub Page and its repo</li> </ol> <p>The first step is to set up the main GitHub Page site and the repository that will host it. This is very simple, as you will only need to follow <a href=https://pages.github.com/ >these steps</a>. After you have created the *organization/username*github.io, it is time to configure your <code>Project</code> repository webpage using MkDocs!</p> <ol> <li>Start a new project from cookiecutter or use one from the previous exercise.</li> </ol> <p>If you use a <code>Project</code> repo from the first exercise, go to the next paragraph. Using cookiecutter, create a new data analysis project. Remember to fill up your metadata and description files! After you have created the folder, it would be best to initialize a Git repo following the instructions from the <a href=#creating-a-git-repo-online-and-copying-your-project-folder>previous section</a>.</p> <p>Next, link your data of interest (or create a small fake dataset) and make an example of data analysis notebook/report (this could be just a scatter plot of a random matrix of values). Depending on your setup, you might be using Jupyter Notebooks or Rmarkdowns. The extensions that we have installed using <code>pip</code> allows you to directly add a Jupyter Notebook file to the <code>mkdocs.yml</code> navigation section. On the other hand, if you are using Rmarkdown, you will have to knit your document into either an html page or a github document.</p> <p>For the purposes of this exercise, we have already included a basic <code>index.md</code> markdown file that can serve as the intro page of your repo, and a <code>jupyter_example.ipynb</code> with some code in it. You are welcome to modify them further to test them out!</p> <ol> <li>Use MkDocs to create your webpage</li> </ol> <p>When you are happy with your files and are ready too publish them, make sure to add, commit and push the changes to the remote. Then, build up your webpage using MkDocs and the <a href=https://www.mkdocs.org/user-guide/deploying-your-docs/ ><code>mkdocs gh-deploy</code></a> command from the same directory where the <code>mkdocs.yml</code> file is. For example, if your <code>mkdocs.yml</code> for your <code>Project</code> folder is in <code>/Users/JARH/Projects/project1_JARH_20231010/mkdocs.yml</code>, do <code>cd /Users/JARH/Projects/project1_JARH_20231010/</code> and then <code>mkdocs gh-deploy</code>.</p> <p>Finally, we only need to set up the GitHub <code>Project</code> repo settings.</p> <ol> <li>Publishing your GitHub Page</li> </ol> <p>Go to your GitHub repo settings and configure the Page section. Since you are using the <code>mkdocs gh-deploy</code> command to publish your site in the <code>gh-pages</code> branch (as explained the the mkdocs documentation), we need to change where GitHub is fetching the website from:</p> <p><img alt="GitHub Pages setup" src=images/git_pages.png></p> <ul> <li>Branch should be <code>gh-pages</code></li> <li>Folder should be <code>root</code></li> </ul> <p>After a couple of minutes, your webpage should be ready!</p> </div> <h2 id=6-archive-github-repositories-on-zenodo>6. Archive GitHub repositories on Zenodo<a class=headerlink href=#6-archive-github-repositories-on-zenodo title="Permanent link">¶</a></h2> <p>Archives are dedicated digital platforms designed for the secure storage, curation, and dissemination of scientific data. These repositories hold great importance in the research community as they serve as reliable archives for preserving valuable datasets. Their standardized formats and robust curation processes ensure the long-term accessibility and citability of research findings. Researchers worldwide rely on these repositories to share, discover, and validate scientific information, thereby fostering transparency, collaboration, and the advancement of knowledge across various domains of study.</p> <p>The next practical exercise will be to archive your <code>Project</code> folder that contains the data analyses performed on your NGS data in a repository like Zenodo. We can do this by linking your Zenodo account to your GitHub account.</p> <div class="admonition warning"> <p class=admonition-title>Archiving your NGS data</p> <p>In this practical lesson, we will only archive our data analyses in the <code>Project</code> folders. Your actual NGS data should be deposited in a domain-specific archive such as <a href=https://www.ncbi.nlm.nih.gov/geo/ >Gene Expression Omnibus (GEO)</a> or <a href=https://www.ebi.ac.uk/fg/annotare/login/ >Annotare</a>. If you want to know more about these archives, check out this <a href=10_repos.html>lesson</a></p> </div> <h3 id=zenodo>Zenodo<a class=headerlink href=#zenodo title="Permanent link">¶</a></h3> <p>Zenodo[https://zenodo.org/] is an open-access digital repository designed to facilitate the archiving of scientific research outputs. It operates under the umbrella of the European Organization for Nuclear Research (CERN) and is supported by the European Commission. Zenodo accommodates a broad spectrum of research outputs, including datasets, papers, software, and multimedia files. This versatility makes it an invaluable resource for researchers across a wide array of domains, promoting transparency, collaboration, and the advancement of knowledge on a global scale.</p> <p>Operating on a user-friendly web platform, Zenodo allows researchers to easily upload, share, and preserve their research data and related materials. Upon deposit, each item is assigned a unique Digital Object Identifier (DOI), granting it a citable status and ensuring its long-term accessibility. Additionally, Zenodo provides robust metadata capabilities, enabling researchers to enrich their submissions with detailed contextual information. In addition, it allows you to <a href=https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content>link you GitHub account</a>, providing a streamlined way to archive a specific release of your GitHub repository directly into Zenodo. This integration simplifies the process of preserving a snapshot of your project's progress for long-term accessibility and citation.</p> <div class="admonition question"> <p class=admonition-title>Exercise 6: Archive a <code>Project</code> GitHub repo in Zenodo</p> <ol> <li>In order to archive your GitHub repos in Zenodo, you will first need to <a href=https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content>link your Zenodo and GitHub accounts</a>.</li> <li>Once your accounts are linked, go to your Zenodo GitHub account settings and turn on the GitHub repository you want to archive. <img alt=zenodo_github_link src=images/zenodo_github.png></li> <li>Creating a Zenodo archive is now as simple as <a href=https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository>making a release</a> in your GitHub repository. Remember to make a proper tag! <strong>NOTE: If you make a release before enabling the GitHub repository in Zenodo, it will not appear in Zenodo!</strong> <img alt=github_release src=images/github_release.png></li> <li>Zenodo will automatically detect the release and it should appear in your Zenodo upload page. <img alt=zenodo_archives src=images/zenodo_archives.png></li> <li>This archive is assigned a unique Digital Object Identifier (DOI), making it a citable reference for your work. <img alt=zenodo_example src=images/zenodo_example.png></li> </ol> <p>Before submitting your work in a journal, make sure to link your data analysis repository to <a href=https://zenodo.org/ >Zenodo</a>, get a DOI and cite it in your manuscript!</p> </div> <h2 id=wrap-up>Wrap up<a class=headerlink href=#wrap-up title="Permanent link">¶</a></h2> <p>In this small workshop we have learned how improve the FAIRability of your data, as well as organizing and structuring it in a way that will be much more useful in the future. This advantages do not serve yourself only, but your teammates, group leader and the general scientific population! We hope that you found this workshop useful. If you would like to leave us some comments or suggestions, feel free to answer this form!</p> <p><a href=https://forms.office.com/e/BZkpzDKL0L><strong>FEEDBACK FORM</strong></a></p></div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 30, 2023</span> <br> Created: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 30, 2023</span> </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=10_repos.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Archiving your data"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Archiving your data </div> </div> </a> <a href=keywords.html class="md-footer__link md-footer__link--next" aria-label="Next: Keyword index"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Keyword index </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.cd18aaf1.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>